{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/xiaoyuzhao/Desktop/Paper_writing_XYZ_20200908/Final_Scripts_and_Supplementary_materials_oncogenic_GI_screenings/5_Combinatorial_CROP_seq/CROP_Python'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "pd.options.display.float_format = '{:.4f}'.format\n",
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib inline\n",
    "\n",
    "from perturbseq import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# I. Load CROPseq data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Create Ppop class with four major attributes: self.matrix, self.normalized_matrix, self.cells, self.genes based on sequencing data (matrix.mtx; barcodes.tsv; features.tsv) and cell.identities.csv**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading digital expression data: ./sequencing/outs_S1/outs/filtered_gene_bc_matrices_mex/GRCh38/matrix.mtx...\n",
      "Densifying matrix...\n",
      "Loading guide identities:./sequencing/outs_S1/outs/cell_identities.csv...\n",
      "Generating summary statistics...\n",
      "Done.\n",
      "Loading digital expression data: ./sequencing/outs_S2/outs/filtered_gene_bc_matrices_mex/GRCh38/matrix.mtx...\n",
      "Densifying matrix...\n",
      "Loading guide identities:./sequencing/outs_S2/outs/cell_identities.csv...\n",
      "Generating summary statistics...\n",
      "Done.\n",
      "Loading digital expression data: ./sequencing/outs_S3/outs/filtered_gene_bc_matrices_mex/GRCh38/matrix.mtx...\n",
      "Densifying matrix...\n",
      "Loading guide identities:./sequencing/outs_S3/outs/cell_identities.csv...\n",
      "Generating summary statistics...\n",
      "Done.\n",
      "Loading digital expression data: ./sequencing/outs_S4/outs/filtered_gene_bc_matrices_mex/GRCh38/matrix.mtx...\n",
      "Densifying matrix...\n",
      "Loading guide identities:./sequencing/outs_S4/outs/cell_identities.csv...\n",
      "Generating summary statistics...\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "Ppop1 = CellPopulation.from_file('./sequencing/outs_S1/', filtered=True)\n",
    "Ppop2 = CellPopulation.from_file('./sequencing/outs_S2/', filtered=True)\n",
    "Ppop3 = CellPopulation.from_file('./sequencing/outs_S3/', filtered=True)\n",
    "Ppop4 = CellPopulation.from_file('./sequencing/outs_S4/', filtered=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>genre</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>cv</th>\n",
       "      <th>fano</th>\n",
       "      <th>num_cells_exisit</th>\n",
       "      <th>in_matrix</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>0.0007</td>\n",
       "      <td>0.0268</td>\n",
       "      <td>37.3251</td>\n",
       "      <td>0.9994</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ENSG00000239945</td>\n",
       "      <td>AL627309.3</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ENSG00000277856</td>\n",
       "      <td>AC233755.2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ENSG00000275063</td>\n",
       "      <td>AC233755.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>AC240274.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>0.0377</td>\n",
       "      <td>0.1966</td>\n",
       "      <td>5.2193</td>\n",
       "      <td>1.0260</td>\n",
       "      <td>305</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ENSG00000277475</td>\n",
       "      <td>AC213203.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ENSG00000268674</td>\n",
       "      <td>FAM231C</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33538 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_name            genre   mean    std      cv   fano  \\\n",
       "gene_id                                                                      \n",
       "ENSG00000243485  MIR1302-2HG  Gene Expression 0.0000 0.0000     nan    nan   \n",
       "ENSG00000237613      FAM138A  Gene Expression 0.0000 0.0000     nan    nan   \n",
       "ENSG00000186092        OR4F5  Gene Expression 0.0000 0.0000     nan    nan   \n",
       "ENSG00000238009   AL627309.1  Gene Expression 0.0007 0.0268 37.3251 0.9994   \n",
       "ENSG00000239945   AL627309.3  Gene Expression 0.0000 0.0000     nan    nan   \n",
       "...                      ...              ...    ...    ...     ...    ...   \n",
       "ENSG00000277856   AC233755.2  Gene Expression 0.0000 0.0000     nan    nan   \n",
       "ENSG00000275063   AC233755.1  Gene Expression 0.0000 0.0000     nan    nan   \n",
       "ENSG00000271254   AC240274.1  Gene Expression 0.0377 0.1966  5.2193 1.0260   \n",
       "ENSG00000277475   AC213203.1  Gene Expression 0.0000 0.0000     nan    nan   \n",
       "ENSG00000268674      FAM231C  Gene Expression 0.0000 0.0000     nan    nan   \n",
       "\n",
       "                 num_cells_exisit  in_matrix  \n",
       "gene_id                                       \n",
       "ENSG00000243485                 0       True  \n",
       "ENSG00000237613                 0       True  \n",
       "ENSG00000186092                 0       True  \n",
       "ENSG00000238009                 6       True  \n",
       "ENSG00000239945                 0       True  \n",
       "...                           ...        ...  \n",
       "ENSG00000277856                 0       True  \n",
       "ENSG00000275063                 0       True  \n",
       "ENSG00000271254               305       True  \n",
       "ENSG00000277475                 0       True  \n",
       "ENSG00000268674                 0       True  \n",
       "\n",
       "[33538 rows x 8 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##Use pop4 as an example\n",
    "Ppop4.genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Condition</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>num_guides</th>\n",
       "      <th>guide_identity</th>\n",
       "      <th>guide_UMI_count</th>\n",
       "      <th>perturbed_1</th>\n",
       "      <th>perturbed_2</th>\n",
       "      <th>num_targets</th>\n",
       "      <th>knock_outs</th>\n",
       "      <th>percent_mito</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>AAACCTGAGAAACGCC-1</td>\n",
       "      <td>S4_D6_1per</td>\n",
       "      <td>5271</td>\n",
       "      <td>1260</td>\n",
       "      <td>2</td>\n",
       "      <td>TP53_4|CTRL0001</td>\n",
       "      <td>13|5</td>\n",
       "      <td>TP53_4-CTRL1</td>\n",
       "      <td>TP53-CTRL1</td>\n",
       "      <td>1</td>\n",
       "      <td>SKO</td>\n",
       "      <td>0.0628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGAGACTTTCG-1</td>\n",
       "      <td>S4_D6_1per</td>\n",
       "      <td>8408</td>\n",
       "      <td>1725</td>\n",
       "      <td>2</td>\n",
       "      <td>NF2_2|CTRL0001</td>\n",
       "      <td>35|25</td>\n",
       "      <td>NF2_2-CTRL1</td>\n",
       "      <td>NF2-CTRL1</td>\n",
       "      <td>1</td>\n",
       "      <td>SKO</td>\n",
       "      <td>0.0258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGAGCTCTCGG-1</td>\n",
       "      <td>S4_D6_1per</td>\n",
       "      <td>11252</td>\n",
       "      <td>2261</td>\n",
       "      <td>2</td>\n",
       "      <td>NF2_2|TBX3_3</td>\n",
       "      <td>43|33</td>\n",
       "      <td>NF2_2-TBX3_3</td>\n",
       "      <td>NF2-TBX3</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGAGGGCTTCC-1</td>\n",
       "      <td>S4_D6_1per</td>\n",
       "      <td>17593</td>\n",
       "      <td>3030</td>\n",
       "      <td>2</td>\n",
       "      <td>PTEN_5|CBFB_2</td>\n",
       "      <td>33|8</td>\n",
       "      <td>PTEN_5-CBFB_2</td>\n",
       "      <td>PTEN-CBFB</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGAGGGTTCCC-1</td>\n",
       "      <td>S4_D6_1per</td>\n",
       "      <td>9658</td>\n",
       "      <td>1538</td>\n",
       "      <td>1</td>\n",
       "      <td>SMAD4_5</td>\n",
       "      <td>51</td>\n",
       "      <td>SMAD4_5-CTRL1</td>\n",
       "      <td>SMAD4-CTRL1</td>\n",
       "      <td>1</td>\n",
       "      <td>SKO</td>\n",
       "      <td>0.0362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGTCAGTTAAGAAC-1</td>\n",
       "      <td>S4_D6_1per</td>\n",
       "      <td>4632</td>\n",
       "      <td>1042</td>\n",
       "      <td>1</td>\n",
       "      <td>RB1_3</td>\n",
       "      <td>4</td>\n",
       "      <td>RB1_3-CTRL1</td>\n",
       "      <td>RB1-CTRL1</td>\n",
       "      <td>1</td>\n",
       "      <td>SKO</td>\n",
       "      <td>0.0484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGTCATCAGAGGTG-1</td>\n",
       "      <td>S4_D6_1per</td>\n",
       "      <td>19258</td>\n",
       "      <td>2870</td>\n",
       "      <td>4</td>\n",
       "      <td>SMAD4_5|CBFB_2|USP9X_3|CTRL0001</td>\n",
       "      <td>49|17|21|19</td>\n",
       "      <td>SMAD4_5-CBFB_2-USP9X_3-CTRL1</td>\n",
       "      <td>SMAD4-CBFB-USP9X-CTRL1</td>\n",
       "      <td>3</td>\n",
       "      <td>MKO</td>\n",
       "      <td>0.0737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGTCATCCCTTGCA-1</td>\n",
       "      <td>S4_D6_1per</td>\n",
       "      <td>22015</td>\n",
       "      <td>3431</td>\n",
       "      <td>2</td>\n",
       "      <td>CASP8_5|CDH1_5</td>\n",
       "      <td>54|21</td>\n",
       "      <td>CASP8_5-CDH1_5</td>\n",
       "      <td>CASP8-CDH1</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGTCATCCGCGCAA-1</td>\n",
       "      <td>S4_D6_1per</td>\n",
       "      <td>15428</td>\n",
       "      <td>2695</td>\n",
       "      <td>3</td>\n",
       "      <td>NF2_2|USP9X_3|RB1_3</td>\n",
       "      <td>26|6|24</td>\n",
       "      <td>NF2_2-USP9X_3-RB1_3</td>\n",
       "      <td>NF2-USP9X-RB1</td>\n",
       "      <td>3</td>\n",
       "      <td>MKO</td>\n",
       "      <td>0.0524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGTCATCTTGTTTG-1</td>\n",
       "      <td>S4_D6_1per</td>\n",
       "      <td>4570</td>\n",
       "      <td>1129</td>\n",
       "      <td>1</td>\n",
       "      <td>PTEN_5</td>\n",
       "      <td>6</td>\n",
       "      <td>PTEN_5-CTRL1</td>\n",
       "      <td>PTEN-CTRL1</td>\n",
       "      <td>1</td>\n",
       "      <td>SKO</td>\n",
       "      <td>0.0438</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8364 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Condition  nCount_RNA  nFeature_RNA  num_guides  \\\n",
       "cell_barcode                                                           \n",
       "AAACCTGAGAAACGCC-1  S4_D6_1per        5271          1260           2   \n",
       "AAACCTGAGACTTTCG-1  S4_D6_1per        8408          1725           2   \n",
       "AAACCTGAGCTCTCGG-1  S4_D6_1per       11252          2261           2   \n",
       "AAACCTGAGGGCTTCC-1  S4_D6_1per       17593          3030           2   \n",
       "AAACCTGAGGGTTCCC-1  S4_D6_1per        9658          1538           1   \n",
       "...                        ...         ...           ...         ...   \n",
       "TTTGTCAGTTAAGAAC-1  S4_D6_1per        4632          1042           1   \n",
       "TTTGTCATCAGAGGTG-1  S4_D6_1per       19258          2870           4   \n",
       "TTTGTCATCCCTTGCA-1  S4_D6_1per       22015          3431           2   \n",
       "TTTGTCATCCGCGCAA-1  S4_D6_1per       15428          2695           3   \n",
       "TTTGTCATCTTGTTTG-1  S4_D6_1per        4570          1129           1   \n",
       "\n",
       "                                     guide_identity guide_UMI_count  \\\n",
       "cell_barcode                                                          \n",
       "AAACCTGAGAAACGCC-1                  TP53_4|CTRL0001            13|5   \n",
       "AAACCTGAGACTTTCG-1                   NF2_2|CTRL0001           35|25   \n",
       "AAACCTGAGCTCTCGG-1                     NF2_2|TBX3_3           43|33   \n",
       "AAACCTGAGGGCTTCC-1                    PTEN_5|CBFB_2            33|8   \n",
       "AAACCTGAGGGTTCCC-1                          SMAD4_5              51   \n",
       "...                                             ...             ...   \n",
       "TTTGTCAGTTAAGAAC-1                            RB1_3               4   \n",
       "TTTGTCATCAGAGGTG-1  SMAD4_5|CBFB_2|USP9X_3|CTRL0001     49|17|21|19   \n",
       "TTTGTCATCCCTTGCA-1                   CASP8_5|CDH1_5           54|21   \n",
       "TTTGTCATCCGCGCAA-1              NF2_2|USP9X_3|RB1_3         26|6|24   \n",
       "TTTGTCATCTTGTTTG-1                           PTEN_5               6   \n",
       "\n",
       "                                     perturbed_1             perturbed_2  \\\n",
       "cell_barcode                                                               \n",
       "AAACCTGAGAAACGCC-1                  TP53_4-CTRL1              TP53-CTRL1   \n",
       "AAACCTGAGACTTTCG-1                   NF2_2-CTRL1               NF2-CTRL1   \n",
       "AAACCTGAGCTCTCGG-1                  NF2_2-TBX3_3                NF2-TBX3   \n",
       "AAACCTGAGGGCTTCC-1                 PTEN_5-CBFB_2               PTEN-CBFB   \n",
       "AAACCTGAGGGTTCCC-1                 SMAD4_5-CTRL1             SMAD4-CTRL1   \n",
       "...                                          ...                     ...   \n",
       "TTTGTCAGTTAAGAAC-1                   RB1_3-CTRL1               RB1-CTRL1   \n",
       "TTTGTCATCAGAGGTG-1  SMAD4_5-CBFB_2-USP9X_3-CTRL1  SMAD4-CBFB-USP9X-CTRL1   \n",
       "TTTGTCATCCCTTGCA-1                CASP8_5-CDH1_5              CASP8-CDH1   \n",
       "TTTGTCATCCGCGCAA-1           NF2_2-USP9X_3-RB1_3           NF2-USP9X-RB1   \n",
       "TTTGTCATCTTGTTTG-1                  PTEN_5-CTRL1              PTEN-CTRL1   \n",
       "\n",
       "                    num_targets knock_outs  percent_mito  \n",
       "cell_barcode                                              \n",
       "AAACCTGAGAAACGCC-1            1        SKO        0.0628  \n",
       "AAACCTGAGACTTTCG-1            1        SKO        0.0258  \n",
       "AAACCTGAGCTCTCGG-1            2        DKO        0.0507  \n",
       "AAACCTGAGGGCTTCC-1            2        DKO        0.0446  \n",
       "AAACCTGAGGGTTCCC-1            1        SKO        0.0362  \n",
       "...                         ...        ...           ...  \n",
       "TTTGTCAGTTAAGAAC-1            1        SKO        0.0484  \n",
       "TTTGTCATCAGAGGTG-1            3        MKO        0.0737  \n",
       "TTTGTCATCCCTTGCA-1            2        DKO        0.0549  \n",
       "TTTGTCATCCGCGCAA-1            3        MKO        0.0524  \n",
       "TTTGTCATCTTGTTTG-1            1        SKO        0.0438  \n",
       "\n",
       "[8364 rows x 11 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##Use pop4 as an example\n",
    "Ppop4.cells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**To be consistent with threshold of Seurat**\n",
    "\n",
    "*Single cells expressing over 800 and less than 5000 nFeature_RNA are kept and multiplets and droplets carring 0 or >2 guides were removed. Genes expressed in >=3 cells were kept.*\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating summary statistics...\n",
      "Done.\n",
      "Generating summary statistics...\n",
      "Done.\n",
      "Generating summary statistics...\n",
      "Done.\n",
      "Generating summary statistics...\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "pop1 = Ppop1.subpopulation(genes = 'num_cells_exisit >= 3', \n",
    "                        cells = '(knock_outs in @cells_interested) and (num_guides in @single_cells) and (nFeature_RNA > 800) & (nFeature_RNA < 5000) & percent_mito < 0.1', \n",
    "                        cells_interested = ['control', 'SKO', 'DKO'],\n",
    "                        single_cells = [1, 2])\n",
    "pop2 = Ppop2.subpopulation(genes = 'num_cells_exisit >= 3', \n",
    "                        cells = '(knock_outs in @cells_interested) and (num_guides in @single_cells) and (nFeature_RNA > 800) & (nFeature_RNA < 5000) & percent_mito < 0.1', \n",
    "                        cells_interested = ['control', 'SKO', 'DKO'],\n",
    "                        single_cells = [1, 2])\n",
    "pop3 = Ppop3.subpopulation(genes = 'num_cells_exisit >= 3', \n",
    "                        cells = '(knock_outs in @cells_interested) and (num_guides in @single_cells) and (nFeature_RNA > 800) & (nFeature_RNA < 5000) & percent_mito < 0.1', \n",
    "                        cells_interested = ['control', 'SKO', 'DKO'],\n",
    "                        single_cells = [1, 2])\n",
    "pop4 = Ppop4.subpopulation(genes = 'num_cells_exisit >= 3', \n",
    "                        cells = '(knock_outs in @cells_interested) and (num_guides in @single_cells) and (nFeature_RNA > 800) & (nFeature_RNA < 5000) & percent_mito < 0.1', \n",
    "                        cells_interested = ['control', 'SKO', 'DKO'],\n",
    "                        single_cells = [1, 2])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Condition</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>num_guides</th>\n",
       "      <th>guide_identity</th>\n",
       "      <th>guide_UMI_count</th>\n",
       "      <th>perturbed_1</th>\n",
       "      <th>perturbed_2</th>\n",
       "      <th>num_targets</th>\n",
       "      <th>knock_outs</th>\n",
       "      <th>percent_mito</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>AAACCTGAGCGTTTAC-1</td>\n",
       "      <td>S2_D6_full</td>\n",
       "      <td>11337</td>\n",
       "      <td>2280</td>\n",
       "      <td>1</td>\n",
       "      <td>SMAD4_5</td>\n",
       "      <td>24</td>\n",
       "      <td>SMAD4_5-CTRL1</td>\n",
       "      <td>SMAD4-CTRL1</td>\n",
       "      <td>1</td>\n",
       "      <td>SKO</td>\n",
       "      <td>0.0436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGAGTGCAAGC-1</td>\n",
       "      <td>S2_D6_full</td>\n",
       "      <td>17230</td>\n",
       "      <td>2360</td>\n",
       "      <td>2</td>\n",
       "      <td>NF2_2|CBFB_2</td>\n",
       "      <td>16|89</td>\n",
       "      <td>NF2_2-CBFB_2</td>\n",
       "      <td>NF2-CBFB</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGCATCGTCGG-1</td>\n",
       "      <td>S2_D6_full</td>\n",
       "      <td>12232</td>\n",
       "      <td>2357</td>\n",
       "      <td>2</td>\n",
       "      <td>SMAD4_5|RB1_3</td>\n",
       "      <td>38|28</td>\n",
       "      <td>SMAD4_5-RB1_3</td>\n",
       "      <td>SMAD4-RB1</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGGTCAACATC-1</td>\n",
       "      <td>S2_D6_full</td>\n",
       "      <td>13711</td>\n",
       "      <td>2111</td>\n",
       "      <td>2</td>\n",
       "      <td>TP53_3|PTEN_5</td>\n",
       "      <td>34|31</td>\n",
       "      <td>TP53_3-PTEN_5</td>\n",
       "      <td>TP53-PTEN</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGGTCGTTGTA-1</td>\n",
       "      <td>S2_D6_full</td>\n",
       "      <td>8662</td>\n",
       "      <td>1551</td>\n",
       "      <td>2</td>\n",
       "      <td>TP53_3|NF1_1</td>\n",
       "      <td>61|37</td>\n",
       "      <td>TP53_3-NF1_1</td>\n",
       "      <td>TP53-NF1</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGTCAGTCGTCTTC-1</td>\n",
       "      <td>S2_D6_full</td>\n",
       "      <td>13001</td>\n",
       "      <td>2348</td>\n",
       "      <td>1</td>\n",
       "      <td>SMAD4_5</td>\n",
       "      <td>38</td>\n",
       "      <td>SMAD4_5-CTRL1</td>\n",
       "      <td>SMAD4-CTRL1</td>\n",
       "      <td>1</td>\n",
       "      <td>SKO</td>\n",
       "      <td>0.0427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGTCAGTTGAACTC-1</td>\n",
       "      <td>S2_D6_full</td>\n",
       "      <td>12036</td>\n",
       "      <td>2274</td>\n",
       "      <td>2</td>\n",
       "      <td>SMAD4_5|RB1_3</td>\n",
       "      <td>33|22</td>\n",
       "      <td>SMAD4_5-RB1_3</td>\n",
       "      <td>SMAD4-RB1</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGTCAGTTTGTTTC-1</td>\n",
       "      <td>S2_D6_full</td>\n",
       "      <td>9364</td>\n",
       "      <td>1793</td>\n",
       "      <td>2</td>\n",
       "      <td>NF2_1|TP53_5</td>\n",
       "      <td>18|39</td>\n",
       "      <td>NF2_1-TP53_5</td>\n",
       "      <td>NF2-TP53</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGTCATCACTTATC-1</td>\n",
       "      <td>S2_D6_full</td>\n",
       "      <td>17549</td>\n",
       "      <td>2487</td>\n",
       "      <td>2</td>\n",
       "      <td>PTEN_5|NF1_1</td>\n",
       "      <td>32|62</td>\n",
       "      <td>PTEN_5-NF1_1</td>\n",
       "      <td>PTEN-NF1</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGTCATCCTTGCCA-1</td>\n",
       "      <td>S2_D6_full</td>\n",
       "      <td>8851</td>\n",
       "      <td>1920</td>\n",
       "      <td>2</td>\n",
       "      <td>NF2_2|CBFB_2</td>\n",
       "      <td>13|15</td>\n",
       "      <td>NF2_2-CBFB_2</td>\n",
       "      <td>NF2-CBFB</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0351</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6402 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Condition  nCount_RNA  nFeature_RNA  num_guides  \\\n",
       "cell_barcode                                                           \n",
       "AAACCTGAGCGTTTAC-1  S2_D6_full       11337          2280           1   \n",
       "AAACCTGAGTGCAAGC-1  S2_D6_full       17230          2360           2   \n",
       "AAACCTGCATCGTCGG-1  S2_D6_full       12232          2357           2   \n",
       "AAACCTGGTCAACATC-1  S2_D6_full       13711          2111           2   \n",
       "AAACCTGGTCGTTGTA-1  S2_D6_full        8662          1551           2   \n",
       "...                        ...         ...           ...         ...   \n",
       "TTTGTCAGTCGTCTTC-1  S2_D6_full       13001          2348           1   \n",
       "TTTGTCAGTTGAACTC-1  S2_D6_full       12036          2274           2   \n",
       "TTTGTCAGTTTGTTTC-1  S2_D6_full        9364          1793           2   \n",
       "TTTGTCATCACTTATC-1  S2_D6_full       17549          2487           2   \n",
       "TTTGTCATCCTTGCCA-1  S2_D6_full        8851          1920           2   \n",
       "\n",
       "                   guide_identity guide_UMI_count    perturbed_1  perturbed_2  \\\n",
       "cell_barcode                                                                    \n",
       "AAACCTGAGCGTTTAC-1        SMAD4_5              24  SMAD4_5-CTRL1  SMAD4-CTRL1   \n",
       "AAACCTGAGTGCAAGC-1   NF2_2|CBFB_2           16|89   NF2_2-CBFB_2     NF2-CBFB   \n",
       "AAACCTGCATCGTCGG-1  SMAD4_5|RB1_3           38|28  SMAD4_5-RB1_3    SMAD4-RB1   \n",
       "AAACCTGGTCAACATC-1  TP53_3|PTEN_5           34|31  TP53_3-PTEN_5    TP53-PTEN   \n",
       "AAACCTGGTCGTTGTA-1   TP53_3|NF1_1           61|37   TP53_3-NF1_1     TP53-NF1   \n",
       "...                           ...             ...            ...          ...   \n",
       "TTTGTCAGTCGTCTTC-1        SMAD4_5              38  SMAD4_5-CTRL1  SMAD4-CTRL1   \n",
       "TTTGTCAGTTGAACTC-1  SMAD4_5|RB1_3           33|22  SMAD4_5-RB1_3    SMAD4-RB1   \n",
       "TTTGTCAGTTTGTTTC-1   NF2_1|TP53_5           18|39   NF2_1-TP53_5     NF2-TP53   \n",
       "TTTGTCATCACTTATC-1   PTEN_5|NF1_1           32|62   PTEN_5-NF1_1     PTEN-NF1   \n",
       "TTTGTCATCCTTGCCA-1   NF2_2|CBFB_2           13|15   NF2_2-CBFB_2     NF2-CBFB   \n",
       "\n",
       "                    num_targets knock_outs  percent_mito  \n",
       "cell_barcode                                              \n",
       "AAACCTGAGCGTTTAC-1            1        SKO        0.0436  \n",
       "AAACCTGAGTGCAAGC-1            2        DKO        0.0353  \n",
       "AAACCTGCATCGTCGG-1            2        DKO        0.0219  \n",
       "AAACCTGGTCAACATC-1            2        DKO        0.0396  \n",
       "AAACCTGGTCGTTGTA-1            2        DKO        0.0344  \n",
       "...                         ...        ...           ...  \n",
       "TTTGTCAGTCGTCTTC-1            1        SKO        0.0427  \n",
       "TTTGTCAGTTGAACTC-1            2        DKO        0.0410  \n",
       "TTTGTCAGTTTGTTTC-1            2        DKO        0.0399  \n",
       "TTTGTCATCACTTATC-1            2        DKO        0.0549  \n",
       "TTTGTCATCCTTGCCA-1            2        DKO        0.0351  \n",
       "\n",
       "[6402 rows x 11 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pop2.cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matrix\n",
      "==========================\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 4976 entries, AAACCTGAGGCATTGG-1 to TTTGTCATCTACTTAC-1\n",
      "Columns: 16735 entries, ENSG00000238009 to ENSG00000271254\n",
      "dtypes: int32(16735)\n",
      "memory usage: 318.0 MB\n"
     ]
    }
   ],
   "source": [
    "##Use pop4 as an example\n",
    "pop1.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# II. Add fitness to metadata of cells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**1. No fitness data for pop1.cells at Day0**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "    perturbed_1: perturbations at sgRNA level.\n",
    "    perturbed_1_Ab.ordered: same with perturbed_1, except ordering two perturbations of DKOs alphabetically.\n",
    "\n",
    "    perturbed_2: pertubations at gene level.\n",
    "    \n",
    "    perturbed: same with perturbed_2, except replacing ('CTRL2-CTRL1', 'CTRL1-CTRL1', 'hRosa26-CTRL1') with \n",
    "    'control', and ordering two perturbations of DKOs alphabetically.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "76\n",
      "67\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['CDH1_5-USP9X_3', 'CBFB_2-CTRL1', 'NF2_2-TP53_4', 'CBFB_2-CDH1_5',\n",
       "       'RB1_3-CTRL1', 'CDH1_5-CTRL1', 'CASP8_5-CTRL1', 'RB1_3-TP53_3',\n",
       "       'SMAD4_5-CTRL1', 'PTEN_5-CTRL1', 'RB1_3-SMAD4_5', 'PTEN_5-SMAD4_5',\n",
       "       'PTEN_5-TP53_3', 'TBX3_3-CTRL1', 'PTEN_5-RB1_3', 'NF2_2-SMAD4_5',\n",
       "       'USP9X_3-CTRL1', 'CASP8_5-NF2_2', 'NF1_1-CTRL1', 'TP53_3-CTRL1',\n",
       "       'NF1_1-SMAD4_5', 'CBFB_2-SMAD4_5', 'CASP8_5-RB1_3', 'NF2_1-TP53_5',\n",
       "       'NF1_1-TP53_3', 'RB1_3-USP9X_3', 'NF2_2-USP9X_3', 'CASP8_5-NF1_1',\n",
       "       'SMAD4_5-TBX3_3', 'SMAD4_5-TP53_3', 'NF2_2-TP53_3',\n",
       "       'CDH1_5-SMAD4_5', 'TP53_5-CTRL1', 'CASP8_5-TBX3_3',\n",
       "       'NF1_1-USP9X_3', 'NF2_2-CTRL1', 'hRosa26_2-CTRL1',\n",
       "       'CASP8_5-SMAD4_5', 'SMAD4_5-USP9X_3', 'TBX3_3-TP53_3',\n",
       "       'CASP8_5-TP53_3', 'CASP8_5-CDH1_5', 'CDH1_5-PTEN_5',\n",
       "       'NF1_1-TBX3_3', 'CDH1_5-NF1_1', 'TP53_4-CTRL1', 'RB1_3-TBX3_3',\n",
       "       'CTRL1-CTRL1', 'NF2_2-TBX3_3', 'NF2_2-PTEN_5', 'NF1_1-RB1_3',\n",
       "       'CTRL2-CTRL1', 'PTEN_5-USP9X_3', 'NF1_1-PTEN_5', 'CBFB_2-PTEN_5',\n",
       "       'CBFB_2-TP53_3', 'CBFB_2-NF2_2', 'CASP8_5-PTEN_5', 'CDH1_5-TP53_3',\n",
       "       'CASP8_5-USP9X_3', 'CBFB_2-TBX3_3', 'CDH1_5-RB1_3', 'CBFB_2-RB1_3',\n",
       "       'PTEN_5-TBX3_3', 'TP53_3-USP9X_3', 'TBX3_3-USP9X_3',\n",
       "       'CDH1_5-TBX3_3', 'NF2_2-RB1_3', 'CBFB_2-NF1_1', 'CDH1_5-NF2_2',\n",
       "       'NF1_1-NF2_2', 'CBFB_2-USP9X_3', 'CBFB_2-TP53_5', 'NF2_1-CTRL1',\n",
       "       'PTEN_5-CTRL2', 'RB1_3-TP53_4'], dtype=object)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Adjust names of pertubations in pop1.cells\n",
    "pop1.cells['perturbed_1'] = pop1.cells['perturbed_1'].map(lambda x: 'RB1_3-CTRL1' if x =='CTRL1-RB1_3' else x )\n",
    "pop1.cells['perturbed_2'] = pop1.cells['perturbed_2'].map(lambda x: 'RB1-CTRL1' if x =='CTRL1-RB1' else x )\n",
    "control_targets = ['CTRL2-CTRL1', 'CTRL1-CTRL1', 'hRosa26-CTRL1']\n",
    "#control_targets = ['CTRL2-CTRL1', 'hRosa26-CTRL1']\n",
    "pop1.cells['perturbed'] = pop1.cells['perturbed_2'].map(lambda x: 'CTRL-CTRL' if x in control_targets else x)\n",
    "\n",
    "pop1.cells['perturbed_1_Ab.ordered'] = pop1.cells['perturbed_1'].map(lambda x: '-'.join(sorted(x.split('-'))) if 'CTRL' not in x   else x)\n",
    "pop1.cells['perturbed'] = pop1.cells['perturbed'].map(lambda x: '-'.join(sorted(x.split('-'))) if 'CTRL' not in x   else x)\n",
    "\n",
    "print(len(pop1.cells['perturbed_1_Ab.ordered'].unique()))\n",
    "print(len(pop1.cells['perturbed'].unique()))\n",
    "\n",
    "##perturbed_1_Ab.ordered is used for mapping sgRNA-level fitness\n",
    "perturbed_CROPseq = pop1.cells['perturbed_1_Ab.ordered'].unique()\n",
    "perturbed_CROPseq\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**2. Mapping fitness data from CRISPR screening to pop2.cells**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "75\n",
      "67\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['SMAD4_5-CTRL1', 'CBFB_2-NF2_2', 'RB1_3-SMAD4_5', 'PTEN_5-TP53_3',\n",
       "       'NF1_1-TP53_3', 'TBX3_3-TP53_3', 'NF2_2-SMAD4_5', 'RB1_3-CTRL1',\n",
       "       'CBFB_2-CTRL1', 'TP53_5-CTRL1', 'SMAD4_5-TP53_3', 'TP53_3-CTRL1',\n",
       "       'CDH1_5-CTRL1', 'PTEN_5-CTRL1', 'PTEN_5-SMAD4_5', 'RB1_3-USP9X_3',\n",
       "       'TP53_4-CTRL1', 'CBFB_2-USP9X_3', 'CASP8_5-CTRL1',\n",
       "       'SMAD4_5-USP9X_3', 'NF2_2-TBX3_3', 'NF2_1-TP53_5', 'NF2_2-TP53_3',\n",
       "       'CDH1_5-TBX3_3', 'SMAD4_5-TBX3_3', 'NF1_1-CTRL1', 'CDH1_5-TP53_3',\n",
       "       'NF2_2-RB1_3', 'NF2_2-USP9X_3', 'TBX3_3-CTRL1', 'PTEN_5-TBX3_3',\n",
       "       'NF2_2-TP53_4', 'CDH1_5-SMAD4_5', 'NF2_2-CTRL1', 'CDH1_5-PTEN_5',\n",
       "       'NF2_2-PTEN_5', 'NF1_1-RB1_3', 'CDH1_5-NF1_1', 'TP53_3-USP9X_3',\n",
       "       'hRosa26_2-CTRL1', 'NF1_1-SMAD4_5', 'CBFB_2-SMAD4_5',\n",
       "       'CASP8_5-CDH1_5', 'CBFB_2-CDH1_5', 'USP9X_3-CTRL1',\n",
       "       'CASP8_5-SMAD4_5', 'CBFB_2-NF1_1', 'CBFB_2-RB1_3',\n",
       "       'CASP8_5-PTEN_5', 'CASP8_5-NF2_2', 'RB1_3-TBX3_3', 'NF1_1-PTEN_5',\n",
       "       'CDH1_5-RB1_3', 'PTEN_5-RB1_3', 'NF1_1-NF2_2', 'CDH1_5-NF2_2',\n",
       "       'CASP8_5-RB1_3', 'CTRL2-CTRL1', 'CDH1_5-USP9X_3', 'CASP8_5-TP53_3',\n",
       "       'CBFB_2-TP53_3', 'NF1_1-TBX3_3', 'CTRL1-CTRL1', 'CBFB_2-PTEN_5',\n",
       "       'CASP8_5-TBX3_3', 'CASP8_5-NF1_1', 'RB1_3-TP53_3',\n",
       "       'CASP8_5-USP9X_3', 'CBFB_2-TBX3_3', 'TBX3_3-USP9X_3',\n",
       "       'PTEN_5-USP9X_3', 'NF2_1-CTRL1', 'NF1_1-USP9X_3', 'PTEN_5-TP53_5',\n",
       "       'TP53_3-TP53_4'], dtype=object)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Adjust names of pertubations in pop2.cells\n",
    "pop2.cells['perturbed_1'] = pop2.cells['perturbed_1'].map(lambda x: 'RB1_3-CTRL1' if x =='CTRL1-RB1_3' else x )\n",
    "pop2.cells['perturbed_2'] = pop2.cells['perturbed_2'].map(lambda x: 'RB1-CTRL1' if x =='CTRL1-RB1' else x )\n",
    "control_targets = ['CTRL2-CTRL1', 'CTRL1-CTRL1', 'hRosa26-CTRL1']\n",
    "#control_targets = ['CTRL2-CTRL1', 'hRosa26-CTRL1']\n",
    "pop2.cells['perturbed'] = pop2.cells['perturbed_2'].map(lambda x: 'CTRL-CTRL' if x in control_targets else x)\n",
    "\n",
    "pop2.cells['perturbed_1_Ab.ordered'] = pop2.cells['perturbed_1'].map(lambda x: '-'.join(sorted(x.split('-'))) if 'CTRL' not in x   else x)\n",
    "pop2.cells['perturbed'] = pop2.cells['perturbed'].map(lambda x: '-'.join(sorted(x.split('-'))) if 'CTRL' not in x   else x)\n",
    "\n",
    "print(len(pop2.cells['perturbed_1_Ab.ordered'].unique()))\n",
    "print(len(pop2.cells['perturbed'].unique()))\n",
    "\n",
    "##perturbed_1_Ab.ordered is used for mapping sgRNA-level fitness\n",
    "perturbed_CROPseq = pop2.cells['perturbed_1_Ab.ordered'].unique()\n",
    "perturbed_CROPseq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape - fitness replicates 1: (33071, 9)\n",
      "Shape - fitness replicates 2: (33115, 9)\n",
      "Shape - fitness replicates 2: (33023, 3)\n",
      "75\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fitness1</th>\n",
       "      <th>fitness2</th>\n",
       "      <th>fitness</th>\n",
       "      <th>Condition</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>num_guides</th>\n",
       "      <th>guide_identity</th>\n",
       "      <th>guide_UMI_count</th>\n",
       "      <th>perturbed_1</th>\n",
       "      <th>perturbed_2</th>\n",
       "      <th>num_targets</th>\n",
       "      <th>knock_outs</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>perturbed</th>\n",
       "      <th>perturbed_1_Ab.ordered</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>AAACCTGAGCGTTTAC-1</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>0.0526</td>\n",
       "      <td>0.0323</td>\n",
       "      <td>S2_D6_full</td>\n",
       "      <td>11337</td>\n",
       "      <td>2280</td>\n",
       "      <td>1</td>\n",
       "      <td>SMAD4_5</td>\n",
       "      <td>24</td>\n",
       "      <td>SMAD4_5-CTRL1</td>\n",
       "      <td>SMAD4-CTRL1</td>\n",
       "      <td>1</td>\n",
       "      <td>SKO</td>\n",
       "      <td>0.0436</td>\n",
       "      <td>SMAD4-CTRL1</td>\n",
       "      <td>SMAD4_5-CTRL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGAGTGCAAGC-1</td>\n",
       "      <td>-0.0117</td>\n",
       "      <td>-0.0603</td>\n",
       "      <td>-0.0360</td>\n",
       "      <td>S2_D6_full</td>\n",
       "      <td>17230</td>\n",
       "      <td>2360</td>\n",
       "      <td>2</td>\n",
       "      <td>NF2_2|CBFB_2</td>\n",
       "      <td>16|89</td>\n",
       "      <td>NF2_2-CBFB_2</td>\n",
       "      <td>NF2-CBFB</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0353</td>\n",
       "      <td>CBFB-NF2</td>\n",
       "      <td>CBFB_2-NF2_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGCATCGTCGG-1</td>\n",
       "      <td>0.0553</td>\n",
       "      <td>0.0548</td>\n",
       "      <td>0.0551</td>\n",
       "      <td>S2_D6_full</td>\n",
       "      <td>12232</td>\n",
       "      <td>2357</td>\n",
       "      <td>2</td>\n",
       "      <td>SMAD4_5|RB1_3</td>\n",
       "      <td>38|28</td>\n",
       "      <td>SMAD4_5-RB1_3</td>\n",
       "      <td>SMAD4-RB1</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0219</td>\n",
       "      <td>RB1-SMAD4</td>\n",
       "      <td>RB1_3-SMAD4_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGGTCAACATC-1</td>\n",
       "      <td>0.1244</td>\n",
       "      <td>0.1136</td>\n",
       "      <td>0.1190</td>\n",
       "      <td>S2_D6_full</td>\n",
       "      <td>13711</td>\n",
       "      <td>2111</td>\n",
       "      <td>2</td>\n",
       "      <td>TP53_3|PTEN_5</td>\n",
       "      <td>34|31</td>\n",
       "      <td>TP53_3-PTEN_5</td>\n",
       "      <td>TP53-PTEN</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0396</td>\n",
       "      <td>PTEN-TP53</td>\n",
       "      <td>PTEN_5-TP53_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGGTCGTTGTA-1</td>\n",
       "      <td>0.0377</td>\n",
       "      <td>0.0448</td>\n",
       "      <td>0.0412</td>\n",
       "      <td>S2_D6_full</td>\n",
       "      <td>8662</td>\n",
       "      <td>1551</td>\n",
       "      <td>2</td>\n",
       "      <td>TP53_3|NF1_1</td>\n",
       "      <td>61|37</td>\n",
       "      <td>TP53_3-NF1_1</td>\n",
       "      <td>TP53-NF1</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0344</td>\n",
       "      <td>NF1-TP53</td>\n",
       "      <td>NF1_1-TP53_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGTCAGTCGTCTTC-1</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>0.0526</td>\n",
       "      <td>0.0323</td>\n",
       "      <td>S2_D6_full</td>\n",
       "      <td>13001</td>\n",
       "      <td>2348</td>\n",
       "      <td>1</td>\n",
       "      <td>SMAD4_5</td>\n",
       "      <td>38</td>\n",
       "      <td>SMAD4_5-CTRL1</td>\n",
       "      <td>SMAD4-CTRL1</td>\n",
       "      <td>1</td>\n",
       "      <td>SKO</td>\n",
       "      <td>0.0427</td>\n",
       "      <td>SMAD4-CTRL1</td>\n",
       "      <td>SMAD4_5-CTRL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGTCAGTTGAACTC-1</td>\n",
       "      <td>0.0553</td>\n",
       "      <td>0.0548</td>\n",
       "      <td>0.0551</td>\n",
       "      <td>S2_D6_full</td>\n",
       "      <td>12036</td>\n",
       "      <td>2274</td>\n",
       "      <td>2</td>\n",
       "      <td>SMAD4_5|RB1_3</td>\n",
       "      <td>33|22</td>\n",
       "      <td>SMAD4_5-RB1_3</td>\n",
       "      <td>SMAD4-RB1</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0410</td>\n",
       "      <td>RB1-SMAD4</td>\n",
       "      <td>RB1_3-SMAD4_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGTCAGTTTGTTTC-1</td>\n",
       "      <td>0.0725</td>\n",
       "      <td>0.0602</td>\n",
       "      <td>0.0664</td>\n",
       "      <td>S2_D6_full</td>\n",
       "      <td>9364</td>\n",
       "      <td>1793</td>\n",
       "      <td>2</td>\n",
       "      <td>NF2_1|TP53_5</td>\n",
       "      <td>18|39</td>\n",
       "      <td>NF2_1-TP53_5</td>\n",
       "      <td>NF2-TP53</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0399</td>\n",
       "      <td>NF2-TP53</td>\n",
       "      <td>NF2_1-TP53_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGTCATCACTTATC-1</td>\n",
       "      <td>0.0196</td>\n",
       "      <td>-0.0137</td>\n",
       "      <td>0.0030</td>\n",
       "      <td>S2_D6_full</td>\n",
       "      <td>17549</td>\n",
       "      <td>2487</td>\n",
       "      <td>2</td>\n",
       "      <td>PTEN_5|NF1_1</td>\n",
       "      <td>32|62</td>\n",
       "      <td>PTEN_5-NF1_1</td>\n",
       "      <td>PTEN-NF1</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0549</td>\n",
       "      <td>NF1-PTEN</td>\n",
       "      <td>NF1_1-PTEN_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGTCATCCTTGCCA-1</td>\n",
       "      <td>-0.0117</td>\n",
       "      <td>-0.0603</td>\n",
       "      <td>-0.0360</td>\n",
       "      <td>S2_D6_full</td>\n",
       "      <td>8851</td>\n",
       "      <td>1920</td>\n",
       "      <td>2</td>\n",
       "      <td>NF2_2|CBFB_2</td>\n",
       "      <td>13|15</td>\n",
       "      <td>NF2_2-CBFB_2</td>\n",
       "      <td>NF2-CBFB</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0351</td>\n",
       "      <td>CBFB-NF2</td>\n",
       "      <td>CBFB_2-NF2_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6402 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    fitness1  fitness2  fitness   Condition  nCount_RNA  \\\n",
       "cell_barcode                                                              \n",
       "AAACCTGAGCGTTTAC-1    0.0120    0.0526   0.0323  S2_D6_full       11337   \n",
       "AAACCTGAGTGCAAGC-1   -0.0117   -0.0603  -0.0360  S2_D6_full       17230   \n",
       "AAACCTGCATCGTCGG-1    0.0553    0.0548   0.0551  S2_D6_full       12232   \n",
       "AAACCTGGTCAACATC-1    0.1244    0.1136   0.1190  S2_D6_full       13711   \n",
       "AAACCTGGTCGTTGTA-1    0.0377    0.0448   0.0412  S2_D6_full        8662   \n",
       "...                      ...       ...      ...         ...         ...   \n",
       "TTTGTCAGTCGTCTTC-1    0.0120    0.0526   0.0323  S2_D6_full       13001   \n",
       "TTTGTCAGTTGAACTC-1    0.0553    0.0548   0.0551  S2_D6_full       12036   \n",
       "TTTGTCAGTTTGTTTC-1    0.0725    0.0602   0.0664  S2_D6_full        9364   \n",
       "TTTGTCATCACTTATC-1    0.0196   -0.0137   0.0030  S2_D6_full       17549   \n",
       "TTTGTCATCCTTGCCA-1   -0.0117   -0.0603  -0.0360  S2_D6_full        8851   \n",
       "\n",
       "                    nFeature_RNA  num_guides guide_identity guide_UMI_count  \\\n",
       "cell_barcode                                                                  \n",
       "AAACCTGAGCGTTTAC-1          2280           1        SMAD4_5              24   \n",
       "AAACCTGAGTGCAAGC-1          2360           2   NF2_2|CBFB_2           16|89   \n",
       "AAACCTGCATCGTCGG-1          2357           2  SMAD4_5|RB1_3           38|28   \n",
       "AAACCTGGTCAACATC-1          2111           2  TP53_3|PTEN_5           34|31   \n",
       "AAACCTGGTCGTTGTA-1          1551           2   TP53_3|NF1_1           61|37   \n",
       "...                          ...         ...            ...             ...   \n",
       "TTTGTCAGTCGTCTTC-1          2348           1        SMAD4_5              38   \n",
       "TTTGTCAGTTGAACTC-1          2274           2  SMAD4_5|RB1_3           33|22   \n",
       "TTTGTCAGTTTGTTTC-1          1793           2   NF2_1|TP53_5           18|39   \n",
       "TTTGTCATCACTTATC-1          2487           2   PTEN_5|NF1_1           32|62   \n",
       "TTTGTCATCCTTGCCA-1          1920           2   NF2_2|CBFB_2           13|15   \n",
       "\n",
       "                      perturbed_1  perturbed_2  num_targets knock_outs  \\\n",
       "cell_barcode                                                             \n",
       "AAACCTGAGCGTTTAC-1  SMAD4_5-CTRL1  SMAD4-CTRL1            1        SKO   \n",
       "AAACCTGAGTGCAAGC-1   NF2_2-CBFB_2     NF2-CBFB            2        DKO   \n",
       "AAACCTGCATCGTCGG-1  SMAD4_5-RB1_3    SMAD4-RB1            2        DKO   \n",
       "AAACCTGGTCAACATC-1  TP53_3-PTEN_5    TP53-PTEN            2        DKO   \n",
       "AAACCTGGTCGTTGTA-1   TP53_3-NF1_1     TP53-NF1            2        DKO   \n",
       "...                           ...          ...          ...        ...   \n",
       "TTTGTCAGTCGTCTTC-1  SMAD4_5-CTRL1  SMAD4-CTRL1            1        SKO   \n",
       "TTTGTCAGTTGAACTC-1  SMAD4_5-RB1_3    SMAD4-RB1            2        DKO   \n",
       "TTTGTCAGTTTGTTTC-1   NF2_1-TP53_5     NF2-TP53            2        DKO   \n",
       "TTTGTCATCACTTATC-1   PTEN_5-NF1_1     PTEN-NF1            2        DKO   \n",
       "TTTGTCATCCTTGCCA-1   NF2_2-CBFB_2     NF2-CBFB            2        DKO   \n",
       "\n",
       "                    percent_mito    perturbed perturbed_1_Ab.ordered  \n",
       "cell_barcode                                                          \n",
       "AAACCTGAGCGTTTAC-1        0.0436  SMAD4-CTRL1          SMAD4_5-CTRL1  \n",
       "AAACCTGAGTGCAAGC-1        0.0353     CBFB-NF2           CBFB_2-NF2_2  \n",
       "AAACCTGCATCGTCGG-1        0.0219    RB1-SMAD4          RB1_3-SMAD4_5  \n",
       "AAACCTGGTCAACATC-1        0.0396    PTEN-TP53          PTEN_5-TP53_3  \n",
       "AAACCTGGTCGTTGTA-1        0.0344     NF1-TP53           NF1_1-TP53_3  \n",
       "...                          ...          ...                    ...  \n",
       "TTTGTCAGTCGTCTTC-1        0.0427  SMAD4-CTRL1          SMAD4_5-CTRL1  \n",
       "TTTGTCAGTTGAACTC-1        0.0410    RB1-SMAD4          RB1_3-SMAD4_5  \n",
       "TTTGTCAGTTTGTTTC-1        0.0399     NF2-TP53           NF2_1-TP53_5  \n",
       "TTTGTCATCACTTATC-1        0.0549     NF1-PTEN           NF1_1-PTEN_5  \n",
       "TTTGTCATCCTTGCCA-1        0.0351     CBFB-NF2           CBFB_2-NF2_2  \n",
       "\n",
       "[6402 rows x 16 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Load sgRNA_level fitness data\n",
    "fit1 = pd.read_csv('./sequencing/outs_S2/outs/full.1.log2RPM_fitlm_20200224.csv', index_col=0)\n",
    "fit2 = pd.read_csv('./sequencing/outs_S2/outs/full.2.log2RPM_fitlm_20200224.csv', index_col=0)\n",
    "print(\"Shape - fitness replicates 1:\", fit1.shape)\n",
    "print(\"Shape - fitness replicates 2:\", fit2.shape)\n",
    "\n",
    "fit1.rename({'x': 'fitness1'}, axis=1, inplace=True)\n",
    "fit2.rename({'x': 'fitness2'}, axis=1, inplace=True)\n",
    "\n",
    "#Calculate average fitness from two replicate experiments\n",
    "fit = pd.merge(fit1[['fitness1']], fit2[['fitness2']], left_index = True, right_index = True)\n",
    "fit['fitness'] = fit.mean(axis = 1, skipna = True)\n",
    "fit.reset_index(inplace = True)\n",
    "fit['index'] = fit['index'].map(lambda x: x.replace('CTRL000', 'CTRL'))\n",
    "fit.set_index('index', inplace = True)\n",
    "print(\"Shape - fitness replicates 2:\", fit.shape)\n",
    "fit.sort_values('fitness', ascending = False)\n",
    "\n",
    "#Mapping average sgRNA-level fitness to pop2.cells according the names of guides\n",
    "pop2.cells = pd.merge(fit, pop2.cells, left_index = True, right_on = 'perturbed_1_Ab.ordered', how = 'right')\n",
    "print(len(pop2.cells['perturbed_1_Ab.ordered'].unique()))\n",
    "perturbed_found = pop2.cells['perturbed_1_Ab.ordered'].unique()\n",
    "perturbed_found\n",
    "#np.setdiff1d(perturbed_CROPseq, perturbed_found)\n",
    "pop2.cells.sort_values('fitness', ascending = False)\n",
    "pop2.cells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**3. Mapping fitness data from CRISPR screening to pop3.cells**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "74\n",
      "66\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['CASP8_5-SMAD4_5', 'CDH1_5-CTRL1', 'SMAD4_5-CTRL1', 'CBFB_2-CTRL1',\n",
       "       'CASP8_5-TP53_3', 'SMAD4_5-TP53_3', 'SMAD4_5-TBX3_3',\n",
       "       'NF1_1-CTRL1', 'TP53_3-CTRL1', 'RB1_3-CTRL1', 'TP53_3-USP9X_3',\n",
       "       'RB1_3-SMAD4_5', 'CDH1_5-SMAD4_5', 'TP53_5-CTRL1', 'NF2_2-SMAD4_5',\n",
       "       'NF2_2-TBX3_3', 'NF2_2-RB1_3', 'NF2_1-TP53_5', 'NF2_2-TP53_4',\n",
       "       'NF2_2-TP53_3', 'CBFB_2-SMAD4_5', 'CBFB_2-NF2_2', 'PTEN_5-RB1_3',\n",
       "       'TBX3_3-TP53_3', 'CASP8_5-CTRL1', 'PTEN_5-TP53_3', 'NF2_2-CTRL1',\n",
       "       'PTEN_5-CTRL1', 'CTRL1-CTRL1', 'CASP8_5-CDH1_5', 'CDH1_5-RB1_3',\n",
       "       'CBFB_2-NF1_1', 'TP53_4-CTRL1', 'NF1_1-SMAD4_5', 'NF2_2-USP9X_3',\n",
       "       'hRosa26_2-CTRL1', 'SMAD4_5-USP9X_3', 'CBFB_2-CDH1_5',\n",
       "       'CBFB_2-TBX3_3', 'PTEN_5-SMAD4_5', 'NF1_1-NF2_2', 'CDH1_5-NF2_2',\n",
       "       'CDH1_5-TP53_3', 'CASP8_5-USP9X_3', 'NF1_1-RB1_3', 'CDH1_5-TBX3_3',\n",
       "       'CASP8_5-RB1_3', 'TBX3_3-CTRL1', 'CBFB_2-RB1_3', 'NF2_2-PTEN_5',\n",
       "       'CASP8_5-NF2_2', 'TBX3_3-USP9X_3', 'RB1_3-TP53_3', 'CDH1_5-NF1_1',\n",
       "       'NF1_1-TP53_3', 'CASP8_5-PTEN_5', 'CASP8_5-TBX3_3', 'NF1_1-PTEN_5',\n",
       "       'RB1_3-TBX3_3', 'CTRL2-CTRL1', 'CDH1_5-PTEN_5', 'USP9X_3-CTRL1',\n",
       "       'PTEN_5-TBX3_3', 'RB1_3-USP9X_3', 'CBFB_2-TP53_3', 'NF1_1-TBX3_3',\n",
       "       'CBFB_2-USP9X_3', 'CDH1_5-USP9X_3', 'CBFB_2-PTEN_5',\n",
       "       'CASP8_5-NF1_1', 'NF1_1-USP9X_3', 'PTEN_5-USP9X_3', 'NF2_1-CTRL1',\n",
       "       'CASP8_5-TP53_4'], dtype=object)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Adjust names of pertubations in pop3.cells\n",
    "pop3.cells['perturbed_1'] = pop3.cells['perturbed_1'].map(lambda x: 'RB1_3-CTRL1' if x =='CTRL1-RB1_3' else x )\n",
    "pop3.cells['perturbed_2'] = pop3.cells['perturbed_2'].map(lambda x: 'RB1-CTRL1' if x =='CTRL1-RB1' else x )\n",
    "control_targets = ['CTRL2-CTRL1', 'CTRL1-CTRL1', 'hRosa26-CTRL1']\n",
    "#control_targets = ['CTRL2-CTRL1', 'hRosa26-CTRL1']\n",
    "pop3.cells['perturbed'] = pop3.cells['perturbed_2'].map(lambda x: 'CTRL-CTRL' if x in control_targets else x)\n",
    "\n",
    "pop3.cells['perturbed_1_Ab.ordered'] = pop3.cells['perturbed_1'].map(lambda x: '-'.join(sorted(x.split('-'))) if 'CTRL' not in x   else x)\n",
    "pop3.cells['perturbed'] = pop3.cells['perturbed'].map(lambda x: '-'.join(sorted(x.split('-'))) if 'CTRL' not in x   else x)\n",
    "\n",
    "print(len(pop3.cells['perturbed_1_Ab.ordered'].unique()))\n",
    "print(len(pop3.cells['perturbed'].unique()))\n",
    "\n",
    "##perturbed_1_Ab.ordered is used for mapping sgRNA-level fitness\n",
    "perturbed_CROPseq = pop3.cells['perturbed_1_Ab.ordered'].unique()\n",
    "perturbed_CROPseq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape - fitness replicates 1: (32673, 9)\n",
      "Shape - fitness replicates 2: (32149, 9)\n",
      "Shape - fitness replicates 2: (32096, 3)\n",
      "74\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Condition</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>num_guides</th>\n",
       "      <th>guide_identity</th>\n",
       "      <th>guide_UMI_count</th>\n",
       "      <th>perturbed_1</th>\n",
       "      <th>perturbed_2</th>\n",
       "      <th>num_targets</th>\n",
       "      <th>knock_outs</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>perturbed</th>\n",
       "      <th>perturbed_1_Ab.ordered</th>\n",
       "      <th>fitness1</th>\n",
       "      <th>fitness2</th>\n",
       "      <th>fitness</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>CCCTCCTAGTGGGATC-1</td>\n",
       "      <td>S3_D6_tgf</td>\n",
       "      <td>18564</td>\n",
       "      <td>2837</td>\n",
       "      <td>2</td>\n",
       "      <td>SMAD4_5|PTEN_5</td>\n",
       "      <td>22|38</td>\n",
       "      <td>SMAD4_5-PTEN_5</td>\n",
       "      <td>SMAD4-PTEN</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0237</td>\n",
       "      <td>PTEN-SMAD4</td>\n",
       "      <td>PTEN_5-SMAD4_5</td>\n",
       "      <td>0.3478</td>\n",
       "      <td>0.3318</td>\n",
       "      <td>0.3398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GTCATTTCAAACGCGA-1</td>\n",
       "      <td>S3_D6_tgf</td>\n",
       "      <td>20877</td>\n",
       "      <td>3242</td>\n",
       "      <td>2</td>\n",
       "      <td>SMAD4_5|PTEN_5</td>\n",
       "      <td>42|23</td>\n",
       "      <td>SMAD4_5-PTEN_5</td>\n",
       "      <td>SMAD4-PTEN</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0524</td>\n",
       "      <td>PTEN-SMAD4</td>\n",
       "      <td>PTEN_5-SMAD4_5</td>\n",
       "      <td>0.3478</td>\n",
       "      <td>0.3318</td>\n",
       "      <td>0.3398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ACCGTAACAGGATCGA-1</td>\n",
       "      <td>S3_D6_tgf</td>\n",
       "      <td>18933</td>\n",
       "      <td>2847</td>\n",
       "      <td>2</td>\n",
       "      <td>SMAD4_5|PTEN_5</td>\n",
       "      <td>16|29</td>\n",
       "      <td>SMAD4_5-PTEN_5</td>\n",
       "      <td>SMAD4-PTEN</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0603</td>\n",
       "      <td>PTEN-SMAD4</td>\n",
       "      <td>PTEN_5-SMAD4_5</td>\n",
       "      <td>0.3478</td>\n",
       "      <td>0.3318</td>\n",
       "      <td>0.3398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TCTGGAAAGGACACCA-1</td>\n",
       "      <td>S3_D6_tgf</td>\n",
       "      <td>33740</td>\n",
       "      <td>4269</td>\n",
       "      <td>2</td>\n",
       "      <td>SMAD4_5|PTEN_5</td>\n",
       "      <td>57|100</td>\n",
       "      <td>SMAD4_5-PTEN_5</td>\n",
       "      <td>SMAD4-PTEN</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0297</td>\n",
       "      <td>PTEN-SMAD4</td>\n",
       "      <td>PTEN_5-SMAD4_5</td>\n",
       "      <td>0.3478</td>\n",
       "      <td>0.3318</td>\n",
       "      <td>0.3398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GTCTCGTGTTCAGACT-1</td>\n",
       "      <td>S3_D6_tgf</td>\n",
       "      <td>31791</td>\n",
       "      <td>3990</td>\n",
       "      <td>2</td>\n",
       "      <td>SMAD4_5|PTEN_5</td>\n",
       "      <td>41|39</td>\n",
       "      <td>SMAD4_5-PTEN_5</td>\n",
       "      <td>SMAD4-PTEN</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0393</td>\n",
       "      <td>PTEN-SMAD4</td>\n",
       "      <td>PTEN_5-SMAD4_5</td>\n",
       "      <td>0.3478</td>\n",
       "      <td>0.3318</td>\n",
       "      <td>0.3398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TGGGCGTCAAACGCGA-1</td>\n",
       "      <td>S3_D6_tgf</td>\n",
       "      <td>13916</td>\n",
       "      <td>2338</td>\n",
       "      <td>1</td>\n",
       "      <td>hRosa26_2</td>\n",
       "      <td>13</td>\n",
       "      <td>hRosa26_2-CTRL1</td>\n",
       "      <td>hRosa26-CTRL1</td>\n",
       "      <td>1</td>\n",
       "      <td>control</td>\n",
       "      <td>0.0501</td>\n",
       "      <td>CTRL-CTRL</td>\n",
       "      <td>hRosa26_2-CTRL1</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTCGAAGAGCTCCTTC-1</td>\n",
       "      <td>S3_D6_tgf</td>\n",
       "      <td>4940</td>\n",
       "      <td>1322</td>\n",
       "      <td>1</td>\n",
       "      <td>CTRL0001</td>\n",
       "      <td>5</td>\n",
       "      <td>CTRL1-CTRL1</td>\n",
       "      <td>CTRL1-CTRL1</td>\n",
       "      <td>0</td>\n",
       "      <td>control</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>CTRL-CTRL</td>\n",
       "      <td>CTRL1-CTRL1</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTGACTTAGCGCCTCA-1</td>\n",
       "      <td>S3_D6_tgf</td>\n",
       "      <td>11400</td>\n",
       "      <td>1757</td>\n",
       "      <td>2</td>\n",
       "      <td>hRosa26_2|CTRL0001</td>\n",
       "      <td>12|35</td>\n",
       "      <td>hRosa26_2-CTRL1</td>\n",
       "      <td>hRosa26-CTRL1</td>\n",
       "      <td>1</td>\n",
       "      <td>control</td>\n",
       "      <td>0.0341</td>\n",
       "      <td>CTRL-CTRL</td>\n",
       "      <td>hRosa26_2-CTRL1</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTGGAACTCATTGCGA-1</td>\n",
       "      <td>S3_D6_tgf</td>\n",
       "      <td>11435</td>\n",
       "      <td>2289</td>\n",
       "      <td>2</td>\n",
       "      <td>CTRL0002|CTRL0001</td>\n",
       "      <td>22|9</td>\n",
       "      <td>CTRL2-CTRL1</td>\n",
       "      <td>CTRL2-CTRL1</td>\n",
       "      <td>0</td>\n",
       "      <td>control</td>\n",
       "      <td>0.0268</td>\n",
       "      <td>CTRL-CTRL</td>\n",
       "      <td>CTRL2-CTRL1</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTCCTCTCCAATGGT-1</td>\n",
       "      <td>S3_D6_tgf</td>\n",
       "      <td>9216</td>\n",
       "      <td>1824</td>\n",
       "      <td>1</td>\n",
       "      <td>USP9X_3</td>\n",
       "      <td>41</td>\n",
       "      <td>USP9X_3-CTRL1</td>\n",
       "      <td>USP9X-CTRL1</td>\n",
       "      <td>1</td>\n",
       "      <td>SKO</td>\n",
       "      <td>0.0216</td>\n",
       "      <td>USP9X-CTRL1</td>\n",
       "      <td>USP9X_3-CTRL1</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5637 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Condition  nCount_RNA  nFeature_RNA  num_guides  \\\n",
       "cell_barcode                                                          \n",
       "CCCTCCTAGTGGGATC-1  S3_D6_tgf       18564          2837           2   \n",
       "GTCATTTCAAACGCGA-1  S3_D6_tgf       20877          3242           2   \n",
       "ACCGTAACAGGATCGA-1  S3_D6_tgf       18933          2847           2   \n",
       "TCTGGAAAGGACACCA-1  S3_D6_tgf       33740          4269           2   \n",
       "GTCTCGTGTTCAGACT-1  S3_D6_tgf       31791          3990           2   \n",
       "...                       ...         ...           ...         ...   \n",
       "TGGGCGTCAAACGCGA-1  S3_D6_tgf       13916          2338           1   \n",
       "TTCGAAGAGCTCCTTC-1  S3_D6_tgf        4940          1322           1   \n",
       "TTGACTTAGCGCCTCA-1  S3_D6_tgf       11400          1757           2   \n",
       "TTGGAACTCATTGCGA-1  S3_D6_tgf       11435          2289           2   \n",
       "TTTCCTCTCCAATGGT-1  S3_D6_tgf        9216          1824           1   \n",
       "\n",
       "                        guide_identity guide_UMI_count      perturbed_1  \\\n",
       "cell_barcode                                                              \n",
       "CCCTCCTAGTGGGATC-1      SMAD4_5|PTEN_5           22|38   SMAD4_5-PTEN_5   \n",
       "GTCATTTCAAACGCGA-1      SMAD4_5|PTEN_5           42|23   SMAD4_5-PTEN_5   \n",
       "ACCGTAACAGGATCGA-1      SMAD4_5|PTEN_5           16|29   SMAD4_5-PTEN_5   \n",
       "TCTGGAAAGGACACCA-1      SMAD4_5|PTEN_5          57|100   SMAD4_5-PTEN_5   \n",
       "GTCTCGTGTTCAGACT-1      SMAD4_5|PTEN_5           41|39   SMAD4_5-PTEN_5   \n",
       "...                                ...             ...              ...   \n",
       "TGGGCGTCAAACGCGA-1           hRosa26_2              13  hRosa26_2-CTRL1   \n",
       "TTCGAAGAGCTCCTTC-1            CTRL0001               5      CTRL1-CTRL1   \n",
       "TTGACTTAGCGCCTCA-1  hRosa26_2|CTRL0001           12|35  hRosa26_2-CTRL1   \n",
       "TTGGAACTCATTGCGA-1   CTRL0002|CTRL0001            22|9      CTRL2-CTRL1   \n",
       "TTTCCTCTCCAATGGT-1             USP9X_3              41    USP9X_3-CTRL1   \n",
       "\n",
       "                      perturbed_2  num_targets knock_outs  percent_mito  \\\n",
       "cell_barcode                                                              \n",
       "CCCTCCTAGTGGGATC-1     SMAD4-PTEN            2        DKO        0.0237   \n",
       "GTCATTTCAAACGCGA-1     SMAD4-PTEN            2        DKO        0.0524   \n",
       "ACCGTAACAGGATCGA-1     SMAD4-PTEN            2        DKO        0.0603   \n",
       "TCTGGAAAGGACACCA-1     SMAD4-PTEN            2        DKO        0.0297   \n",
       "GTCTCGTGTTCAGACT-1     SMAD4-PTEN            2        DKO        0.0393   \n",
       "...                           ...          ...        ...           ...   \n",
       "TGGGCGTCAAACGCGA-1  hRosa26-CTRL1            1    control        0.0501   \n",
       "TTCGAAGAGCTCCTTC-1    CTRL1-CTRL1            0    control        0.0103   \n",
       "TTGACTTAGCGCCTCA-1  hRosa26-CTRL1            1    control        0.0341   \n",
       "TTGGAACTCATTGCGA-1    CTRL2-CTRL1            0    control        0.0268   \n",
       "TTTCCTCTCCAATGGT-1    USP9X-CTRL1            1        SKO        0.0216   \n",
       "\n",
       "                      perturbed perturbed_1_Ab.ordered  fitness1  fitness2  \\\n",
       "cell_barcode                                                                 \n",
       "CCCTCCTAGTGGGATC-1   PTEN-SMAD4         PTEN_5-SMAD4_5    0.3478    0.3318   \n",
       "GTCATTTCAAACGCGA-1   PTEN-SMAD4         PTEN_5-SMAD4_5    0.3478    0.3318   \n",
       "ACCGTAACAGGATCGA-1   PTEN-SMAD4         PTEN_5-SMAD4_5    0.3478    0.3318   \n",
       "TCTGGAAAGGACACCA-1   PTEN-SMAD4         PTEN_5-SMAD4_5    0.3478    0.3318   \n",
       "GTCTCGTGTTCAGACT-1   PTEN-SMAD4         PTEN_5-SMAD4_5    0.3478    0.3318   \n",
       "...                         ...                    ...       ...       ...   \n",
       "TGGGCGTCAAACGCGA-1    CTRL-CTRL        hRosa26_2-CTRL1       nan       nan   \n",
       "TTCGAAGAGCTCCTTC-1    CTRL-CTRL            CTRL1-CTRL1       nan       nan   \n",
       "TTGACTTAGCGCCTCA-1    CTRL-CTRL        hRosa26_2-CTRL1       nan       nan   \n",
       "TTGGAACTCATTGCGA-1    CTRL-CTRL            CTRL2-CTRL1       nan       nan   \n",
       "TTTCCTCTCCAATGGT-1  USP9X-CTRL1          USP9X_3-CTRL1       nan       nan   \n",
       "\n",
       "                    fitness  \n",
       "cell_barcode                 \n",
       "CCCTCCTAGTGGGATC-1   0.3398  \n",
       "GTCATTTCAAACGCGA-1   0.3398  \n",
       "ACCGTAACAGGATCGA-1   0.3398  \n",
       "TCTGGAAAGGACACCA-1   0.3398  \n",
       "GTCTCGTGTTCAGACT-1   0.3398  \n",
       "...                     ...  \n",
       "TGGGCGTCAAACGCGA-1      nan  \n",
       "TTCGAAGAGCTCCTTC-1      nan  \n",
       "TTGACTTAGCGCCTCA-1      nan  \n",
       "TTGGAACTCATTGCGA-1      nan  \n",
       "TTTCCTCTCCAATGGT-1      nan  \n",
       "\n",
       "[5637 rows x 16 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fit1 = pd.read_csv('./sequencing/outs_S3/outs/tgf.1.log2RPM_fitlm_20200224.csv', index_col=0)\n",
    "fit2 = pd.read_csv('./sequencing/outs_S3/outs/tgf.2.log2RPM_fitlm_20200224.csv', index_col=0)\n",
    "print(\"Shape - fitness replicates 1:\", fit1.shape)\n",
    "print(\"Shape - fitness replicates 2:\", fit2.shape)\n",
    "\n",
    "fit1.rename({'x': 'fitness1'}, axis=1, inplace=True)\n",
    "fit2.rename({'x': 'fitness2'}, axis=1, inplace=True)\n",
    "\n",
    "#Calculate average fitness from two replicate experiments\n",
    "fit = pd.merge(fit1[['fitness1']], fit2[['fitness2']], left_index = True, right_index = True)\n",
    "fit['fitness'] = fit.mean(axis = 1, skipna = True)\n",
    "fit.reset_index(inplace = True)\n",
    "fit['index'] = fit['index'].map(lambda x: x.replace('CTRL000', 'CTRL'))\n",
    "fit.set_index('index', inplace = True)\n",
    "print(\"Shape - fitness replicates 2:\", fit.shape)\n",
    "fit.sort_values('fitness', ascending = False)\n",
    "\n",
    "#Mapping average sgRNA-level fitness to pop3.cells according the names of guides\n",
    "pop3.cells = pd.merge(pop3.cells, fit, left_on = 'perturbed_1_Ab.ordered', right_index = True, how = 'left')\n",
    "print(len(pop3.cells['perturbed_1_Ab.ordered'].unique()))\n",
    "perturbed_found = pop3.cells['perturbed_1_Ab.ordered'].unique()\n",
    "perturbed_found\n",
    "#np.setdiff1d(perturbed_CROPseq, perturbed_found)\n",
    "pop3.cells.sort_values('fitness', ascending = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**4. Mapping fitness data from CRISPR screening to pop4.cells**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "80\n",
      "68\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['TP53_4-CTRL1', 'NF2_2-CTRL1', 'NF2_2-TBX3_3', 'CBFB_2-PTEN_5',\n",
       "       'SMAD4_5-CTRL1', 'PTEN_5-CTRL1', 'NF1_1-RB1_3', 'CDH1_5-PTEN_5',\n",
       "       'TP53_5-CTRL1', 'USP9X_3-CTRL1', 'RB1_3-CTRL1', 'CBFB_2-NF2_2',\n",
       "       'PTEN_5-RB1_3', 'TBX3_3-CTRL1', 'NF2_2-TP53_4', 'NF2_1-TP53_5',\n",
       "       'NF1_1-PTEN_5', 'CASP8_5-CTRL1', 'NF2_2-SMAD4_5', 'PTEN_5-TP53_3',\n",
       "       'CASP8_5-SMAD4_5', 'CDH1_5-SMAD4_5', 'CDH1_5-CTRL1',\n",
       "       'SMAD4_5-USP9X_3', 'CBFB_2-CTRL1', 'NF1_1-NF2_2', 'PTEN_5-TBX3_3',\n",
       "       'NF1_1-CTRL1', 'CDH1_5-TBX3_3', 'TP53_3-CTRL1', 'CDH1_5-TP53_3',\n",
       "       'PTEN_5-USP9X_3', 'NF2_2-RB1_3', 'NF1_1-TBX3_3', 'SMAD4_5-TBX3_3',\n",
       "       'CBFB_2-TP53_3', 'SMAD4_5-TP53_3', 'CASP8_5-TP53_3',\n",
       "       'hRosa26_2-CTRL1', 'RB1_3-SMAD4_5', 'CASP8_5-USP9X_3',\n",
       "       'CBFB_2-USP9X_3', 'RB1_3-TBX3_3', 'NF1_1-USP9X_3', 'NF1_1-SMAD4_5',\n",
       "       'PTEN_5-SMAD4_5', 'CASP8_5-RB1_3', 'CTRL2-CTRL1', 'NF1_1-TP53_3',\n",
       "       'NF2_2-TP53_3', 'CBFB_2-SMAD4_5', 'CBFB_2-TBX3_3', 'CDH1_5-NF1_1',\n",
       "       'CBFB_2-CDH1_5', 'RB1_3-TP53_3', 'CBFB_2-RB1_3', 'CASP8_5-NF1_1',\n",
       "       'CBFB_2-NF1_1', 'NF2_2-USP9X_3', 'CASP8_5-TBX3_3', 'CDH1_5-RB1_3',\n",
       "       'RB1_3-USP9X_3', 'CDH1_5-USP9X_3', 'CASP8_5-PTEN_5',\n",
       "       'CASP8_5-CDH1_5', 'TBX3_3-TP53_3', 'NF2_2-PTEN_5', 'CASP8_5-NF2_2',\n",
       "       'CDH1_5-NF2_2', 'PTEN_5-TP53_4', 'CTRL1-CTRL1', 'TBX3_3-USP9X_3',\n",
       "       'TP53_3-USP9X_3', 'PTEN_5-hRosa26_2', 'NF2_1-CTRL1',\n",
       "       'NF1_1-TP53_5', 'SMAD4_5-hRosa26_2', 'RB1_3-TP53_5',\n",
       "       'SMAD4_5-TP53_4', 'CASP8_5-NF2_1'], dtype=object)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Adjust names of pertubations in pop4.cells\n",
    "pop4.cells['perturbed_1'] = pop4.cells['perturbed_1'].map(lambda x: 'RB1_3-CTRL1' if x =='CTRL1-RB1_3' else x )\n",
    "pop4.cells['perturbed_2'] = pop4.cells['perturbed_2'].map(lambda x: 'RB1-CTRL1' if x =='CTRL1-RB1' else x )\n",
    "control_targets = ['CTRL2-CTRL1', 'CTRL1-CTRL1', 'hRosa26-CTRL1']\n",
    "#control_targets = ['CTRL2-CTRL1', 'hRosa26-CTRL1']\n",
    "pop4.cells['perturbed'] = pop4.cells['perturbed_2'].map(lambda x: 'CTRL-CTRL' if x in control_targets else x)\n",
    "\n",
    "pop4.cells['perturbed_1_Ab.ordered'] = pop4.cells['perturbed_1'].map(lambda x: '-'.join(sorted(x.split('-'))) if 'CTRL' not in x   else x)\n",
    "pop4.cells['perturbed'] = pop4.cells['perturbed'].map(lambda x: '-'.join(sorted(x.split('-'))) if 'CTRL' not in x   else x)\n",
    "\n",
    "print(len(pop4.cells['perturbed_1_Ab.ordered'].unique()))\n",
    "print(len(pop4.cells['perturbed'].unique()))\n",
    "\n",
    "##perturbed_1_Ab.ordered is used for mapping sgRNA-level fitness\n",
    "perturbed_CROPseq = pop4.cells['perturbed_1_Ab.ordered'].unique()\n",
    "perturbed_CROPseq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape - fitness replicates 1: (33070, 9)\n",
      "Shape - fitness replicates 2: (33116, 9)\n",
      "Shape - fitness replicates 2: (33023, 3)\n",
      "80\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Condition</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>num_guides</th>\n",
       "      <th>guide_identity</th>\n",
       "      <th>guide_UMI_count</th>\n",
       "      <th>perturbed_1</th>\n",
       "      <th>perturbed_2</th>\n",
       "      <th>num_targets</th>\n",
       "      <th>knock_outs</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>perturbed</th>\n",
       "      <th>perturbed_1_Ab.ordered</th>\n",
       "      <th>fitness1</th>\n",
       "      <th>fitness2</th>\n",
       "      <th>fitness</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ATCCACCAGAGAACAG-1</td>\n",
       "      <td>S4_D6_1per</td>\n",
       "      <td>14886</td>\n",
       "      <td>2234</td>\n",
       "      <td>2</td>\n",
       "      <td>NF2_2|PTEN_5</td>\n",
       "      <td>19|26</td>\n",
       "      <td>NF2_2-PTEN_5</td>\n",
       "      <td>NF2-PTEN</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0650</td>\n",
       "      <td>NF2-PTEN</td>\n",
       "      <td>NF2_2-PTEN_5</td>\n",
       "      <td>0.3752</td>\n",
       "      <td>0.3873</td>\n",
       "      <td>0.3813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ACACTGAAGTTACCCA-1</td>\n",
       "      <td>S4_D6_1per</td>\n",
       "      <td>11323</td>\n",
       "      <td>2422</td>\n",
       "      <td>2</td>\n",
       "      <td>NF2_2|PTEN_5</td>\n",
       "      <td>24|43</td>\n",
       "      <td>NF2_2-PTEN_5</td>\n",
       "      <td>NF2-PTEN</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0359</td>\n",
       "      <td>NF2-PTEN</td>\n",
       "      <td>NF2_2-PTEN_5</td>\n",
       "      <td>0.3752</td>\n",
       "      <td>0.3873</td>\n",
       "      <td>0.3813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CACACAAGTCTCCCTA-1</td>\n",
       "      <td>S4_D6_1per</td>\n",
       "      <td>11123</td>\n",
       "      <td>2552</td>\n",
       "      <td>2</td>\n",
       "      <td>NF2_2|PTEN_5</td>\n",
       "      <td>18|27</td>\n",
       "      <td>NF2_2-PTEN_5</td>\n",
       "      <td>NF2-PTEN</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0439</td>\n",
       "      <td>NF2-PTEN</td>\n",
       "      <td>NF2_2-PTEN_5</td>\n",
       "      <td>0.3752</td>\n",
       "      <td>0.3873</td>\n",
       "      <td>0.3813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ACGCAGCTCGTCTGCT-1</td>\n",
       "      <td>S4_D6_1per</td>\n",
       "      <td>15344</td>\n",
       "      <td>2452</td>\n",
       "      <td>2</td>\n",
       "      <td>NF2_2|PTEN_5</td>\n",
       "      <td>14|49</td>\n",
       "      <td>NF2_2-PTEN_5</td>\n",
       "      <td>NF2-PTEN</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0804</td>\n",
       "      <td>NF2-PTEN</td>\n",
       "      <td>NF2_2-PTEN_5</td>\n",
       "      <td>0.3752</td>\n",
       "      <td>0.3873</td>\n",
       "      <td>0.3813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CCATGTCCATCGGAAG-1</td>\n",
       "      <td>S4_D6_1per</td>\n",
       "      <td>10091</td>\n",
       "      <td>2460</td>\n",
       "      <td>2</td>\n",
       "      <td>NF2_2|PTEN_5</td>\n",
       "      <td>11|12</td>\n",
       "      <td>NF2_2-PTEN_5</td>\n",
       "      <td>NF2-PTEN</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0506</td>\n",
       "      <td>NF2-PTEN</td>\n",
       "      <td>NF2_2-PTEN_5</td>\n",
       "      <td>0.3752</td>\n",
       "      <td>0.3873</td>\n",
       "      <td>0.3813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTGACTTGTATATGAG-1</td>\n",
       "      <td>S4_D6_1per</td>\n",
       "      <td>6660</td>\n",
       "      <td>1450</td>\n",
       "      <td>2</td>\n",
       "      <td>SMAD4_5|TBX3_3</td>\n",
       "      <td>15|8</td>\n",
       "      <td>SMAD4_5-TBX3_3</td>\n",
       "      <td>SMAD4-TBX3</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0895</td>\n",
       "      <td>SMAD4-TBX3</td>\n",
       "      <td>SMAD4_5-TBX3_3</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTGGAACCAGATGGCA-1</td>\n",
       "      <td>S4_D6_1per</td>\n",
       "      <td>10081</td>\n",
       "      <td>1889</td>\n",
       "      <td>2</td>\n",
       "      <td>CTRL0002|CTRL0001</td>\n",
       "      <td>30|7</td>\n",
       "      <td>CTRL2-CTRL1</td>\n",
       "      <td>CTRL2-CTRL1</td>\n",
       "      <td>0</td>\n",
       "      <td>control</td>\n",
       "      <td>0.0559</td>\n",
       "      <td>CTRL-CTRL</td>\n",
       "      <td>CTRL2-CTRL1</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTGGCAAGTTAGATGA-1</td>\n",
       "      <td>S4_D6_1per</td>\n",
       "      <td>11701</td>\n",
       "      <td>2132</td>\n",
       "      <td>2</td>\n",
       "      <td>hRosa26_2|CTRL0001</td>\n",
       "      <td>21|9</td>\n",
       "      <td>hRosa26_2-CTRL1</td>\n",
       "      <td>hRosa26-CTRL1</td>\n",
       "      <td>1</td>\n",
       "      <td>control</td>\n",
       "      <td>0.0373</td>\n",
       "      <td>CTRL-CTRL</td>\n",
       "      <td>hRosa26_2-CTRL1</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGCGCCACGCCAGT-1</td>\n",
       "      <td>S4_D6_1per</td>\n",
       "      <td>11993</td>\n",
       "      <td>1978</td>\n",
       "      <td>2</td>\n",
       "      <td>hRosa26_2|CTRL0001</td>\n",
       "      <td>27|10</td>\n",
       "      <td>hRosa26_2-CTRL1</td>\n",
       "      <td>hRosa26-CTRL1</td>\n",
       "      <td>1</td>\n",
       "      <td>control</td>\n",
       "      <td>0.0626</td>\n",
       "      <td>CTRL-CTRL</td>\n",
       "      <td>hRosa26_2-CTRL1</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGCGCGTCTCTCGT-1</td>\n",
       "      <td>S4_D6_1per</td>\n",
       "      <td>14341</td>\n",
       "      <td>1941</td>\n",
       "      <td>2</td>\n",
       "      <td>SMAD4_5|TBX3_3</td>\n",
       "      <td>18|12</td>\n",
       "      <td>SMAD4_5-TBX3_3</td>\n",
       "      <td>SMAD4-TBX3</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>0.0538</td>\n",
       "      <td>SMAD4-TBX3</td>\n",
       "      <td>SMAD4_5-TBX3_3</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6935 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Condition  nCount_RNA  nFeature_RNA  num_guides  \\\n",
       "cell_barcode                                                           \n",
       "ATCCACCAGAGAACAG-1  S4_D6_1per       14886          2234           2   \n",
       "ACACTGAAGTTACCCA-1  S4_D6_1per       11323          2422           2   \n",
       "CACACAAGTCTCCCTA-1  S4_D6_1per       11123          2552           2   \n",
       "ACGCAGCTCGTCTGCT-1  S4_D6_1per       15344          2452           2   \n",
       "CCATGTCCATCGGAAG-1  S4_D6_1per       10091          2460           2   \n",
       "...                        ...         ...           ...         ...   \n",
       "TTGACTTGTATATGAG-1  S4_D6_1per        6660          1450           2   \n",
       "TTGGAACCAGATGGCA-1  S4_D6_1per       10081          1889           2   \n",
       "TTGGCAAGTTAGATGA-1  S4_D6_1per       11701          2132           2   \n",
       "TTTGCGCCACGCCAGT-1  S4_D6_1per       11993          1978           2   \n",
       "TTTGCGCGTCTCTCGT-1  S4_D6_1per       14341          1941           2   \n",
       "\n",
       "                        guide_identity guide_UMI_count      perturbed_1  \\\n",
       "cell_barcode                                                              \n",
       "ATCCACCAGAGAACAG-1        NF2_2|PTEN_5           19|26     NF2_2-PTEN_5   \n",
       "ACACTGAAGTTACCCA-1        NF2_2|PTEN_5           24|43     NF2_2-PTEN_5   \n",
       "CACACAAGTCTCCCTA-1        NF2_2|PTEN_5           18|27     NF2_2-PTEN_5   \n",
       "ACGCAGCTCGTCTGCT-1        NF2_2|PTEN_5           14|49     NF2_2-PTEN_5   \n",
       "CCATGTCCATCGGAAG-1        NF2_2|PTEN_5           11|12     NF2_2-PTEN_5   \n",
       "...                                ...             ...              ...   \n",
       "TTGACTTGTATATGAG-1      SMAD4_5|TBX3_3            15|8   SMAD4_5-TBX3_3   \n",
       "TTGGAACCAGATGGCA-1   CTRL0002|CTRL0001            30|7      CTRL2-CTRL1   \n",
       "TTGGCAAGTTAGATGA-1  hRosa26_2|CTRL0001            21|9  hRosa26_2-CTRL1   \n",
       "TTTGCGCCACGCCAGT-1  hRosa26_2|CTRL0001           27|10  hRosa26_2-CTRL1   \n",
       "TTTGCGCGTCTCTCGT-1      SMAD4_5|TBX3_3           18|12   SMAD4_5-TBX3_3   \n",
       "\n",
       "                      perturbed_2  num_targets knock_outs  percent_mito  \\\n",
       "cell_barcode                                                              \n",
       "ATCCACCAGAGAACAG-1       NF2-PTEN            2        DKO        0.0650   \n",
       "ACACTGAAGTTACCCA-1       NF2-PTEN            2        DKO        0.0359   \n",
       "CACACAAGTCTCCCTA-1       NF2-PTEN            2        DKO        0.0439   \n",
       "ACGCAGCTCGTCTGCT-1       NF2-PTEN            2        DKO        0.0804   \n",
       "CCATGTCCATCGGAAG-1       NF2-PTEN            2        DKO        0.0506   \n",
       "...                           ...          ...        ...           ...   \n",
       "TTGACTTGTATATGAG-1     SMAD4-TBX3            2        DKO        0.0895   \n",
       "TTGGAACCAGATGGCA-1    CTRL2-CTRL1            0    control        0.0559   \n",
       "TTGGCAAGTTAGATGA-1  hRosa26-CTRL1            1    control        0.0373   \n",
       "TTTGCGCCACGCCAGT-1  hRosa26-CTRL1            1    control        0.0626   \n",
       "TTTGCGCGTCTCTCGT-1     SMAD4-TBX3            2        DKO        0.0538   \n",
       "\n",
       "                     perturbed perturbed_1_Ab.ordered  fitness1  fitness2  \\\n",
       "cell_barcode                                                                \n",
       "ATCCACCAGAGAACAG-1    NF2-PTEN           NF2_2-PTEN_5    0.3752    0.3873   \n",
       "ACACTGAAGTTACCCA-1    NF2-PTEN           NF2_2-PTEN_5    0.3752    0.3873   \n",
       "CACACAAGTCTCCCTA-1    NF2-PTEN           NF2_2-PTEN_5    0.3752    0.3873   \n",
       "ACGCAGCTCGTCTGCT-1    NF2-PTEN           NF2_2-PTEN_5    0.3752    0.3873   \n",
       "CCATGTCCATCGGAAG-1    NF2-PTEN           NF2_2-PTEN_5    0.3752    0.3873   \n",
       "...                        ...                    ...       ...       ...   \n",
       "TTGACTTGTATATGAG-1  SMAD4-TBX3         SMAD4_5-TBX3_3       nan       nan   \n",
       "TTGGAACCAGATGGCA-1   CTRL-CTRL            CTRL2-CTRL1       nan       nan   \n",
       "TTGGCAAGTTAGATGA-1   CTRL-CTRL        hRosa26_2-CTRL1       nan       nan   \n",
       "TTTGCGCCACGCCAGT-1   CTRL-CTRL        hRosa26_2-CTRL1       nan       nan   \n",
       "TTTGCGCGTCTCTCGT-1  SMAD4-TBX3         SMAD4_5-TBX3_3       nan       nan   \n",
       "\n",
       "                    fitness  \n",
       "cell_barcode                 \n",
       "ATCCACCAGAGAACAG-1   0.3813  \n",
       "ACACTGAAGTTACCCA-1   0.3813  \n",
       "CACACAAGTCTCCCTA-1   0.3813  \n",
       "ACGCAGCTCGTCTGCT-1   0.3813  \n",
       "CCATGTCCATCGGAAG-1   0.3813  \n",
       "...                     ...  \n",
       "TTGACTTGTATATGAG-1      nan  \n",
       "TTGGAACCAGATGGCA-1      nan  \n",
       "TTGGCAAGTTAGATGA-1      nan  \n",
       "TTTGCGCCACGCCAGT-1      nan  \n",
       "TTTGCGCGTCTCTCGT-1      nan  \n",
       "\n",
       "[6935 rows x 16 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Load sgRNA_level fitness data\n",
    "fit1 = pd.read_csv('./sequencing/outs_S4/outs/1p1.log2RPM_fitlm_20200224.csv', index_col=0)\n",
    "fit2 = pd.read_csv('./sequencing/outs_S4/outs/1p2.log2RPM_fitlm_20200224.csv', index_col=0)\n",
    "print(\"Shape - fitness replicates 1:\", fit1.shape)\n",
    "print(\"Shape - fitness replicates 2:\", fit2.shape)\n",
    "\n",
    "fit1.rename({'x': 'fitness1'}, axis=1, inplace=True)\n",
    "fit2.rename({'x': 'fitness2'}, axis=1, inplace=True)\n",
    "\n",
    "#Calculate average fitness from two replicate experiments\n",
    "fit = pd.merge(fit1[['fitness1']], fit2[['fitness2']], left_index = True, right_index = True)\n",
    "fit['fitness'] = fit.mean(axis = 1, skipna = True)\n",
    "fit.reset_index(inplace = True)\n",
    "fit['index'] = fit['index'].map(lambda x: x.replace('CTRL000', 'CTRL'))\n",
    "fit.set_index('index', inplace = True)\n",
    "print(\"Shape - fitness replicates 2:\", fit.shape)\n",
    "fit.sort_values('fitness', ascending = False)\n",
    "\n",
    "#Mapping average sgRNA-level fitness to pop4.cells according the names of guides\n",
    "pop4.cells = pd.merge(pop4.cells, fit, left_on = 'perturbed_1_Ab.ordered', right_index = True, how = 'left')\n",
    "print(len(pop4.cells['perturbed_1_Ab.ordered'].unique()))\n",
    "perturbed_found = pop4.cells['perturbed_1_Ab.ordered'].unique()\n",
    "perturbed_found\n",
    "#np.setdiff1d(perturbed_CROPseq, perturbed_found)\n",
    "pop4.cells.sort_values('fitness', ascending = False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# III. Add normalized matrix to the class attribute "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**For memory efficiency, remove genes with no counts from expression matrix.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Optional \n",
    "#strip_low_expression(pop)\n",
    "#pop.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Generate normalized matrix**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Determining scale factors...\n",
      "     Normalizing matrix to median\n",
      "     Normalizing control matrix to median\n",
      "     Scaling matrix to control\n",
      "     Done.\n",
      "     Determining scale factors...\n",
      "     Normalizing matrix to median\n",
      "     Normalizing control matrix to median\n",
      "     Scaling matrix to control\n",
      "     Done.\n",
      "     Determining scale factors...\n",
      "     Normalizing matrix to median\n",
      "     Normalizing control matrix to median\n",
      "     Scaling matrix to control\n",
      "     Done.\n",
      "     Determining scale factors...\n",
      "     Normalizing matrix to median\n",
      "     Normalizing control matrix to median\n",
      "     Scaling matrix to control\n",
      "     Done.\n"
     ]
    }
   ],
   "source": [
    "pop1.normalized_matrix = normalize_to_control(pop1, control_cells='perturbed == \"CTRL-CTRL\"')\n",
    "pop2.normalized_matrix = normalize_to_control(pop2, control_cells='perturbed == \"CTRL-CTRL\"')\n",
    "pop3.normalized_matrix = normalize_to_control(pop3, control_cells='perturbed == \"CTRL-CTRL\"')\n",
    "pop4.normalized_matrix = normalize_to_control(pop4, control_cells='perturbed == \"CTRL-CTRL\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Use pop4 as an example\n",
    "#pop4.normalized_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Save as hdf5 so that we don't have to do slow steps every time notebook runs**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing matrix...\n",
      "Writing normalized matrix...\n",
      "Writing metadata...\n",
      "Done in 7.1548638343811035s.\n",
      "Writing matrix...\n",
      "Writing normalized matrix...\n",
      "Writing metadata...\n",
      "Done in 8.140849828720093s.\n",
      "Writing matrix...\n",
      "Writing normalized matrix...\n",
      "Writing metadata...\n",
      "Done in 6.7508180141448975s.\n",
      "Writing matrix...\n",
      "Writing normalized matrix...\n",
      "Writing metadata...\n",
      "Done in 10.658375978469849s.\n"
     ]
    }
   ],
   "source": [
    "pop1.to_hdf('./sequencing/outs_S1/final_pop_with_normalized_S1.hdf', store_normalized_matrix=True)\n",
    "pop2.to_hdf('./sequencing/outs_S2/final_pop_with_normalized_S2.hdf', store_normalized_matrix=True)\n",
    "pop3.to_hdf('./sequencing/outs_S3/final_pop_with_normalized_S3.hdf', store_normalized_matrix=True)\n",
    "pop4.to_hdf('./sequencing/outs_S4/final_pop_with_normalized_S4.hdf', store_normalized_matrix=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# V. Generate mean population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing average expression matrices...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "18c0f6f82f6941a4b02bcca839a7a3ed",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=1, bar_style='info', max=1), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Computing normalized average expression matrices...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "bb62b350490f4d4696bed6b0cdfc0660",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=1, bar_style='info', max=1), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Computing clusters...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "34b6de7ed4e149aa91d5fb0e340f6607",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=1, bar_style='info', max=1), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Generating summary statistics...\n",
      "Done.\n",
      "Computing average expression matrices...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7731d06a48f042508b9cf0b1986b3d40",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=1, bar_style='info', max=1), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Computing normalized average expression matrices...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "987287b92dcd4eb4af629dcfc4eca535",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=1, bar_style='info', max=1), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Computing clusters...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "94c4f41750454e6ea65a75fdb760345f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=1, bar_style='info', max=1), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Generating summary statistics...\n",
      "Done.\n",
      "Computing average expression matrices...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9bca725ad92240a3b11623b214157fc3",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=1, bar_style='info', max=1), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Computing normalized average expression matrices...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "fd2d9295dac34641a7507aac8c355acf",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=1, bar_style='info', max=1), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Computing clusters...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a534f9b83ae846f0b57c2b2e8ae53706",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=1, bar_style='info', max=1), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Generating summary statistics...\n",
      "Done.\n",
      "Computing average expression matrices...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "84d8584a4cfd484b867b301b506c08da",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=1, bar_style='info', max=1), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Computing normalized average expression matrices...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f6c3f2a84e2e490ba7d16114d9b137f6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=1, bar_style='info', max=1), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Computing clusters...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "89c7bb6fef984501bd927c8b01753669",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(IntProgress(value=1, bar_style='info', max=1), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Generating summary statistics...\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "mean_pop1 = pop1.average('perturbed', show_progress=True)\n",
    "mean_pop2 = pop2.average('perturbed', show_progress=True)\n",
    "mean_pop3 = pop3.average('perturbed', show_progress=True)\n",
    "mean_pop4 = pop4.average('perturbed', show_progress=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**I. Add metadata to mean_pop.cells**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_cells</th>\n",
       "      <th>UMI_mean</th>\n",
       "      <th>num_targets</th>\n",
       "      <th>knock_outs</th>\n",
       "      <th>first_target</th>\n",
       "      <th>second_target</th>\n",
       "      <th>perturbed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>CASP8-CDH1</td>\n",
       "      <td>51</td>\n",
       "      <td>17367.0000</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>CASP8</td>\n",
       "      <td>CDH1</td>\n",
       "      <td>CASP8-CDH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CASP8-CTRL1</td>\n",
       "      <td>178</td>\n",
       "      <td>15557.4157</td>\n",
       "      <td>1</td>\n",
       "      <td>SKO</td>\n",
       "      <td>CASP8</td>\n",
       "      <td>CTRL1</td>\n",
       "      <td>CASP8-CTRL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CASP8-NF1</td>\n",
       "      <td>12</td>\n",
       "      <td>17808.9167</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>CASP8</td>\n",
       "      <td>NF1</td>\n",
       "      <td>CASP8-NF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CASP8-NF2</td>\n",
       "      <td>37</td>\n",
       "      <td>16757.1351</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>CASP8</td>\n",
       "      <td>NF2</td>\n",
       "      <td>CASP8-NF2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CASP8-PTEN</td>\n",
       "      <td>23</td>\n",
       "      <td>17586.1739</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>CASP8</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>CASP8-PTEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TBX3-TP53</td>\n",
       "      <td>31</td>\n",
       "      <td>16211.1290</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>TBX3</td>\n",
       "      <td>TP53</td>\n",
       "      <td>TBX3-TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TBX3-USP9X</td>\n",
       "      <td>26</td>\n",
       "      <td>15523.3462</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>TBX3</td>\n",
       "      <td>USP9X</td>\n",
       "      <td>TBX3-USP9X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TP53-CTRL1</td>\n",
       "      <td>417</td>\n",
       "      <td>16210.0911</td>\n",
       "      <td>1</td>\n",
       "      <td>SKO</td>\n",
       "      <td>TP53</td>\n",
       "      <td>CTRL1</td>\n",
       "      <td>TP53-CTRL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TP53-USP9X</td>\n",
       "      <td>27</td>\n",
       "      <td>17312.6296</td>\n",
       "      <td>2</td>\n",
       "      <td>DKO</td>\n",
       "      <td>TP53</td>\n",
       "      <td>USP9X</td>\n",
       "      <td>TP53-USP9X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>USP9X-CTRL1</td>\n",
       "      <td>55</td>\n",
       "      <td>16153.3636</td>\n",
       "      <td>1</td>\n",
       "      <td>SKO</td>\n",
       "      <td>USP9X</td>\n",
       "      <td>CTRL1</td>\n",
       "      <td>USP9X-CTRL1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             num_cells   UMI_mean  num_targets knock_outs first_target  \\\n",
       "CASP8-CDH1          51 17367.0000            2        DKO        CASP8   \n",
       "CASP8-CTRL1        178 15557.4157            1        SKO        CASP8   \n",
       "CASP8-NF1           12 17808.9167            2        DKO        CASP8   \n",
       "CASP8-NF2           37 16757.1351            2        DKO        CASP8   \n",
       "CASP8-PTEN          23 17586.1739            2        DKO        CASP8   \n",
       "...                ...        ...          ...        ...          ...   \n",
       "TBX3-TP53           31 16211.1290            2        DKO         TBX3   \n",
       "TBX3-USP9X          26 15523.3462            2        DKO         TBX3   \n",
       "TP53-CTRL1         417 16210.0911            1        SKO         TP53   \n",
       "TP53-USP9X          27 17312.6296            2        DKO         TP53   \n",
       "USP9X-CTRL1         55 16153.3636            1        SKO        USP9X   \n",
       "\n",
       "            second_target    perturbed  \n",
       "CASP8-CDH1           CDH1   CASP8-CDH1  \n",
       "CASP8-CTRL1         CTRL1  CASP8-CTRL1  \n",
       "CASP8-NF1             NF1    CASP8-NF1  \n",
       "CASP8-NF2             NF2    CASP8-NF2  \n",
       "CASP8-PTEN           PTEN   CASP8-PTEN  \n",
       "...                   ...          ...  \n",
       "TBX3-TP53            TP53    TBX3-TP53  \n",
       "TBX3-USP9X          USP9X   TBX3-USP9X  \n",
       "TP53-CTRL1          CTRL1   TP53-CTRL1  \n",
       "TP53-USP9X          USP9X   TP53-USP9X  \n",
       "USP9X-CTRL1         CTRL1  USP9X-CTRL1  \n",
       "\n",
       "[67 rows x 7 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mean_pop1.cells = pd.merge(mean_pop1.cells, pop1.cells.groupby('perturbed').mean()[['num_targets']], left_index = True, right_index= True,how = 'left')\n",
    "mean_pop1.cells['num_targets'] = mean_pop1.cells['num_targets'].astype(int)\n",
    "mean_pop1.cells['knock_outs'] = mean_pop1.cells['num_targets'].map(lambda x: 'SKO' if x == 1 else ('DKO' if x == 2 else 'control'))\n",
    "\n",
    "mean_pop1.cells['first_target'] = mean_pop1.cells.index.map(lambda x: x.split('-')[0])\n",
    "mean_pop1.cells['second_target'] = mean_pop1.cells.index.map(lambda x: x.split('-')[1])\n",
    "mean_pop1.cells['perturbed'] = mean_pop1.cells.index\n",
    "\n",
    "mean_pop1.cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_cells</th>\n",
       "      <th>UMI_mean</th>\n",
       "      <th>num_targets</th>\n",
       "      <th>first_target</th>\n",
       "      <th>second_target</th>\n",
       "      <th>perturbed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>knock_outs</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>DKO</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SKO</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>control</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            num_cells  UMI_mean  num_targets  first_target  second_target  \\\n",
       "knock_outs                                                                  \n",
       "DKO                54        54           54            54             54   \n",
       "SKO                12        12           12            12             12   \n",
       "control             1         1            1             1              1   \n",
       "\n",
       "            perturbed  \n",
       "knock_outs             \n",
       "DKO                54  \n",
       "SKO                12  \n",
       "control             1  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mean_pop1.cells.groupby(['knock_outs']).count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_pop2.cells = pd.merge(mean_pop2.cells, pop2.cells.groupby('perturbed').mean()[['num_targets', 'fitness1', 'fitness2', 'fitness']], left_index = True, right_index= True,how = 'left')\n",
    "mean_pop2.cells['num_targets'] = mean_pop2.cells['num_targets'].astype(int)\n",
    "mean_pop2.cells['knock_outs'] = mean_pop2.cells['num_targets'].map(lambda x: 'SKO' if x == 1 else ('DKO' if x == 2 else 'control'))\n",
    "\n",
    "mean_pop2.cells['first_target'] = mean_pop2.cells.index.map(lambda x: x.split('-')[0])\n",
    "mean_pop2.cells['second_target'] = mean_pop2.cells.index.map(lambda x: x.split('-')[1])\n",
    "mean_pop2.cells['perturbed'] = mean_pop2.cells.index\n",
    "\n",
    "#Save mean_pop.cell information\n",
    "#mean_pop2.cells.to_csv('./sequencing/outs_S2/results/mean_pop2.cells.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_cells</th>\n",
       "      <th>UMI_mean</th>\n",
       "      <th>num_targets</th>\n",
       "      <th>fitness1</th>\n",
       "      <th>fitness2</th>\n",
       "      <th>fitness</th>\n",
       "      <th>first_target</th>\n",
       "      <th>second_target</th>\n",
       "      <th>perturbed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>knock_outs</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>DKO</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>53</td>\n",
       "      <td>53</td>\n",
       "      <td>53</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SKO</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>control</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            num_cells  UMI_mean  num_targets  fitness1  fitness2  fitness  \\\n",
       "knock_outs                                                                  \n",
       "DKO                55        55           55        53        53       53   \n",
       "SKO                11        11           11        10        10       10   \n",
       "control             1         1            1         0         0        0   \n",
       "\n",
       "            first_target  second_target  perturbed  \n",
       "knock_outs                                          \n",
       "DKO                   55             55         55  \n",
       "SKO                   11             11         11  \n",
       "control                1              1          1  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mean_pop2.cells.groupby(['knock_outs']).count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_pop3.cells = pd.merge(mean_pop3.cells, pop3.cells.groupby('perturbed').mean()[['num_targets', 'fitness1', 'fitness2', 'fitness']], left_index = True, right_index= True,how = 'left')\n",
    "mean_pop3.cells['num_targets'] = mean_pop3.cells['num_targets'].astype(int)\n",
    "mean_pop3.cells['knock_outs'] = mean_pop3.cells['num_targets'].map(lambda x: 'SKO' if x == 1 else ('DKO' if x == 2 else 'control'))\n",
    "\n",
    "mean_pop3.cells['first_target'] = mean_pop3.cells.index.map(lambda x: x.split('-')[0])\n",
    "mean_pop3.cells['second_target'] = mean_pop3.cells.index.map(lambda x: x.split('-')[1])\n",
    "mean_pop3.cells['perturbed'] = mean_pop3.cells.index\n",
    "\n",
    "#Save mean_pop.cell information\n",
    "#mean_pop3.cells.to_csv('./sequencing/outs_S3/results/mean_pop3.cells.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_cells</th>\n",
       "      <th>UMI_mean</th>\n",
       "      <th>num_targets</th>\n",
       "      <th>fitness1</th>\n",
       "      <th>fitness2</th>\n",
       "      <th>fitness</th>\n",
       "      <th>first_target</th>\n",
       "      <th>second_target</th>\n",
       "      <th>perturbed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>knock_outs</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>DKO</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SKO</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>control</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            num_cells  UMI_mean  num_targets  fitness1  fitness2  fitness  \\\n",
       "knock_outs                                                                  \n",
       "DKO                54        54           54        54        54       54   \n",
       "SKO                11        11           11        10        10       10   \n",
       "control             1         1            1         0         0        0   \n",
       "\n",
       "            first_target  second_target  perturbed  \n",
       "knock_outs                                          \n",
       "DKO                   54             54         54  \n",
       "SKO                   11             11         11  \n",
       "control                1              1          1  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mean_pop3.cells.groupby(['knock_outs']).count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_pop4.cells = pd.merge(mean_pop4.cells, pop4.cells.groupby('perturbed').mean()[['num_targets', 'fitness1', 'fitness2', 'fitness']], left_index = True, right_index= True,how = 'left')\n",
    "mean_pop4.cells['num_targets'] = mean_pop4.cells['num_targets'].astype(int)\n",
    "mean_pop4.cells['knock_outs'] = mean_pop4.cells['num_targets'].map(lambda x: 'SKO' if x == 1 else ('DKO' if x == 2 else 'control'))\n",
    "\n",
    "mean_pop4.cells['first_target'] = mean_pop4.cells.index.map(lambda x: x.split('-')[0])\n",
    "mean_pop4.cells['second_target'] = mean_pop4.cells.index.map(lambda x: x.split('-')[1])\n",
    "mean_pop4.cells['perturbed'] = mean_pop4.cells.index\n",
    "\n",
    "#Save mean_pop.cell information\n",
    "#mean_pop4.cells.to_csv('./sequencing/outs_S4/results/mean_pop4.cells.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_cells</th>\n",
       "      <th>UMI_mean</th>\n",
       "      <th>num_targets</th>\n",
       "      <th>fitness1</th>\n",
       "      <th>fitness2</th>\n",
       "      <th>fitness</th>\n",
       "      <th>first_target</th>\n",
       "      <th>second_target</th>\n",
       "      <th>perturbed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>knock_outs</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>DKO</td>\n",
       "      <td>56</td>\n",
       "      <td>56</td>\n",
       "      <td>56</td>\n",
       "      <td>53</td>\n",
       "      <td>53</td>\n",
       "      <td>53</td>\n",
       "      <td>56</td>\n",
       "      <td>56</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SKO</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>control</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            num_cells  UMI_mean  num_targets  fitness1  fitness2  fitness  \\\n",
       "knock_outs                                                                  \n",
       "DKO                56        56           56        53        53       53   \n",
       "SKO                11        11           11        10        10       10   \n",
       "control             1         1            1         0         0        0   \n",
       "\n",
       "            first_target  second_target  perturbed  \n",
       "knock_outs                                          \n",
       "DKO                   56             56         56  \n",
       "SKO                   11             11         11  \n",
       "control                1              1          1  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mean_pop4.cells.groupby(['knock_outs']).count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Save mean population**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing matrix...\n",
      "Writing normalized matrix...\n",
      "Writing metadata...\n",
      "Done in 0.16013407707214355s.\n",
      "Writing matrix...\n",
      "Writing normalized matrix...\n",
      "Writing metadata...\n",
      "Done in 0.17050695419311523s.\n",
      "Writing matrix...\n",
      "Writing normalized matrix...\n",
      "Writing metadata...\n",
      "Done in 0.14975523948669434s.\n",
      "Writing matrix...\n",
      "Writing normalized matrix...\n",
      "Writing metadata...\n",
      "Done in 0.20610260963439941s.\n"
     ]
    }
   ],
   "source": [
    "mean_pop1.to_hdf('./sequencing/outs_S1/mean_pop_with_normalized_S1.hdf', store_normalized_matrix=True)\n",
    "mean_pop2.to_hdf('./sequencing/outs_S2/mean_pop_with_normalized_S2.hdf', store_normalized_matrix=True)\n",
    "mean_pop3.to_hdf('./sequencing/outs_S3/mean_pop_with_normalized_S3.hdf', store_normalized_matrix=True)\n",
    "mean_pop4.to_hdf('./sequencing/outs_S4/mean_pop_with_normalized_S4.hdf', store_normalized_matrix=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Over!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "combCROPseq",
   "language": "python",
   "name": "combcropseq"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
